---
title: "Manuscript: T Cells"
output:
  html_document:
    code_download: yes
    code_folding: hide
    css: ../"style.css"
    dev: "png"
    df_print: kable
    fig_caption: yes
    keep_md: no
    self_contained: true
    theme: yeti
    toc: yes
    toc_depth: 3
    toc_float:
        collapsed: false
        smooth_scroll: false
editor_options:
  chunk_output_type: console
---

------------------------------------------------------------------------

# Overview and setup

```{r setup-options, include = FALSE}
knitr::opts_chunk$set(comment = NA, dpi = 600, fig.align = "center",
                      out.width = "100%", warning = FALSE, error = TRUE,
                      message = FALSE,
                      fig.path = here::here("primary_analysis", "figures",
                                            "manuscripts", "manuscript-"))

# will mess with chunk labels, but needed for parameterized reports
options(knitr.duplicate.label = "allow")
```

**Author(s):** Edel Aron<br>
**Last Updated:** `r Sys.Date()`

**R version:** `r R.Version()$version.string`<br>
**Platform:** `r R.Version()$platform`<br>
**Running under:** `r sessionInfo()$running`

```{r, include = FALSE}
# load packages, colors and themes and functions (R code only)
code_setup <-
  knitr::purl(here::here("single_cell", "primary_analysis", "notebooks",
                         "01-setup.Rmd"),
              output = tempfile(), documentation = 0)
code_functions <-
  knitr::purl(here::here("single_cell", "primary_analysis", "notebooks",
                         "02-functions.Rmd"),
              output = tempfile(), documentation = 0)
code_meta <- 
  knitr::purl(here::here("single_cell", "primary_analysis", "notebooks",
                         "03-meta.Rmd"),
              output = tempfile(), documentation = 0)

# read in the code, then remove the temp files
source(code_setup)
source(code_functions)
source(code_meta)
unlink(c(code_setup, code_functions, code_meta))
```

Set the figure themes:

```{r}
labels_manuscript <- theme(plot.title = element_text(face = "plain", size = 14,
                                                     hjust = 0.5),
                           plot.subtitle = element_text(size = 12),
                           plot.tag = element_text(face = "plain", size = 12))

labels_manuscript_text <- theme(axis.title.x = element_text(size = 12),
                                axis.title.y = element_text(size = 12),
                                axis.text = element_text(size = 11),
                                legend.title = element_text(size = 12),
                                legend.text = element_text(size = 11),
                                plot.title = element_text(size = 12),
                                plot.subtitle = element_text(size = 11),
                                strip.text.x = element_text(size = 11),
                                strip.text.y = element_text(size = 11))

labels_manuscript_text_big <- theme(axis.title.x = element_text(size = 18),
                                    axis.title.y = element_text(size = 18),
                                    axis.text = element_text(size = 16),
                                    legend.title = element_text(size = 18),
                                    legend.text = element_text(size = 15),
                                    plot.title = element_text(size = 18),
                                    plot.subtitle = element_text(size = 16),
                                    plot.tag =
                                      element_text(face = "plain", size = 18),
                                    strip.text.x = element_text(size = 15),
                                    strip.text.y = element_text(size = 15))

labels_manuscript_dot <- labels_manuscript + labels_manuscript_text +
                         theme(axis.title.x = element_blank(),
                               axis.title.y = element_blank(),
                               # shrink the gap between the facets
                               panel.spacing = unit(0.3, "lines"),
                               strip.text.y = element_text(angle = 0)) +
                         RotatedAxis()

labels_manuscript_prop <- labels_manuscript_text +
                          theme(panel.grid.minor.x = element_blank(),
                                panel.grid.major.y = element_blank(),
                                axis.text.y = element_blank(),
                                axis.ticks.y = element_blank())

labels_manuscript_venn <- labels_manuscript +
                          theme(axis.line = element_blank())

labels_manuscript_violin <- labels_manuscript_text +
                            theme(axis.title.x = element_blank(),
                                  axis.title.y = element_text(size = 11.5))

# update the colors
colors_dataset_manuscript <- setNames(colors_dataset, 1:3)
```

------------------------------------------------------------------------

# Read in the data

Make sure to change the directories as needed based upon your file structure.

## Metadata

```{r}
metrics_summary <- create_metrics_summary(meta = meta,
                                          path_data = path_data_root,
                                          data_types = "GEX",
                                          samples_list = samples)

aggr_results <- fromJSON(file.path(path_data_root, "aggr",
                                   "outs", "count", "summary.json"))
```

```{r}
table_clinical <-
  meta_clinical %>%
    select("Study ID", "Age", "Gender", "Single EM", "Size (diameter in cm)",
           "If yes to skin biopsy obtained from the EM,",
           "If yes to skin biopsy obtained from the EM, location",
           "Dataset", "Paired") %>%
    unite("Lesion Site(s)",
          c("If yes to skin biopsy obtained from the EM,",
            "If yes to skin biopsy obtained from the EM, location"),
          sep = " ") %>%
    rename(Subject = "Study ID", Sex = Gender, EM = "Single EM",
           Size = "Size (diameter in cm)", "Paired Control" = Paired) %>%
    mutate(Subject = as.character(Subject), Size = as.character(Size),
           EM = recode(EM, "Yes" = "Single", "No" = "Multiple"),
           Dataset = as.character(as.integer(Dataset) - 1)) %>%
    relocate(Dataset, .before = Subject)
```

## GEX

### Blood

```{r setup-gex-blood}
# objects
combined_blood <-
  readRDS(file.path(path_objs, "final", "combined_blood_airr.rds"))
combined_blood_t <-
  readRDS(file.path(path_objs, "final", "combined_blood_t_ref.rds"))

# colors
colors_annotated_blood <- combined_blood@misc$colors_annotated
colors_annotated_blood_t <- combined_blood_t@misc$colors_annotated
```

With doublets:

```{r}
combined_blood_with_doublets <-
  readRDS(file.path(path_objs, "doublets", "with_doublets", "combined_blood.rds"))
```

### Skin

```{r setup-gex-skin}
# objects
combined_skin_unprocessed <-
  readRDS(file.path(path_objs, "raw", "combined_skin_unprocessed.rds"))
combined_skin <-
  readRDS(file.path(path_objs, "final", "combined_skin_airr.rds"))
combined_skin_recombined <-
  readRDS(file.path(path_objs, "final", "combined_skin_recombined.rds"))
combined_skin_t <-
  readRDS(file.path(path_objs, "final", "combined_skin_t_ref.rds"))

# for splicing
combined_skin_t_ideis <-
  readRDS(file.path(path_objs, "downstream", "splicing",
                    "combined_skin_t_ideis.rds"))

# colors
colors_annotated_skin <- combined_skin@misc$colors_annotated
colors_annotated_skin_recombined <- combined_skin_recombined@misc$colors_annotated
colors_annotated_skin_t <- combined_skin_t@misc$colors_annotated

colors_cellchat <-
  colors_annotated_skin_recombined[sort(names(colors_annotated_skin_recombined))]
colors_cellchat <- colors_cellchat[!names(colors_cellchat) == "T Cells"]
```

With doublets:

```{r}
combined_skin_with_doublets <-
  readRDS(file.path(path_objs, "doublets", "with_doublets", "combined_skin.rds"))
```

## Differential expression

```{r setup-de}
fdr_cutoff <- 0.05

hallmark <- msigdbr::msigdbr(species = "Homo sapiens", category = "H") %>%
              select(gs_name, gene_symbol)

# this has had the "T Cells" removed
de_skin_recombined <-
  readRDS(file.path(path_objs, "downstream", "differential_expression",
                    "deseq2_combined_skin_recombined.rds"))
de_genes_skin_recombined <- enframe(de_skin_recombined) %>%
                              unnest(value) %>%
                              rename(cell_type = name) %>%
                              dplyr::filter(p_val_adj < 0.05)

de_skin_t_em_ifng <-
  readRDS(file.path(path_objs, "downstream", "differential_expression",
                    "deseq2_combined_skin_t_em_ifng.rds"))
de_genes_skin_t_em_ifng <- enframe(de_skin_t_em_ifng) %>%
                             unnest(value) %>%
                             rename(cell_type = name) %>%
                             dplyr::filter(p_val_adj < 0.05)
```

## CCC

```{r setup-cellchat}
cellchat_list <- readRDS(file.path(path_objs, "downstream", "signaling",
                                   "cellchat_list.rds"))
cellchat_merged <- readRDS(file.path(path_objs, "downstream", "signaling",
                                     "cellchat_merged.rds"))
```

## Other

```{r setup-other}
# for the supported data values spreadsheet later
sdv <- list()
```

------------------------------------------------------------------------

# Update the data

This section of code is for if you are using objects built from the raw data using the other notebooks that we have provided. If you are downloading the objects directly from GEO, you can skip it (and just load those objects in the previous section).

## Dataset

We say datasets 1, 2 and 3 instead of datasets 2, 3 and 4 in the manuscript.

```{r}
# objects
combined_skin$Dataset <-
  as.character(as.integer(combined_skin[[]]$Dataset) - 1)
combined_skin_recombined$Dataset <-
  as.character(as.integer(combined_skin_recombined[[]]$Dataset) - 1)
combined_skin_t$Dataset <-
  as.character(as.integer(combined_skin_t[[]]$Dataset) - 1)

# table
metrics_summary$Dataset <-
  as.character(as.integer(metrics_summary$Dataset) - 1)
```

## Shorter cell type names

To save on room in some of the figures:

```{r}
combined_skin_annotations_shorter <-
  rbind(c("B Cells", "B Cells"),
        c("Cycling Dendritic Cells", "cDCs"),
        c("Endothelial Cells", "Endothelial Cells"),
        c("Fibroblasts", "Fibroblasts"),
        c("Keratinocytes", "Keratinocytes"),
        c("Langerhans", "Langerhans"),
        c("Macrophages", "Macrophages"),
        c("Mast Cells", "Mast Cells"),
        c("Melanocytes", "Melanocytes"),
        c("Natural Killers", "NK Cells"),
        c("Nerves", "Nerves"),
        c("Pericytes", "Pericytes"),
        c("T Cells", "T Cells"))

colnames(combined_skin_annotations_shorter) <- c("Cluster", "CellType")
combined_skin_annotations_shorter <-
  data.frame(combined_skin_annotations_shorter)

combined_skin <-
  add_annotations(seurat_obj = combined_skin,
                  annotations_df = combined_skin_annotations_shorter,
                  relabel = FALSE, clusters_col = "annotated_clusters",
                  annotations_col = "annotated_clusters_shorter")

colors_annotated_skin_shorter <-
  setNames(colors_annotated_skin,
           nm = combined_skin_annotations_shorter$CellType)
```

```{r}
combined_skin_t_annotations_shorter <-
  rbind(c("CD4- CD8- T Cells", "CD4- CD8-"),
        c("CD4+ Effector Memory T Cells", "CD4+ TEMs"),
        c("CD4+ Tissue-Resident Memory T Cells", "CD4+ TRMs"),
        c("CD4+ Undefined T Cells", "CD4+ Undefined"),
        c("CD8+ GZMK+ IFNGhi T Cells", "CD8+ GZMK+ IFNGhi"),
        c("CD8+ GZMK+ IFNGint T Cells", "CD8+ GZMK+ IFNGint"),
        c("CD8+ Undefined T Cells", "CD8+ Undefined"),
        c("Dividing T Cells", "Dividing"),
        c("HSPhi T Cells", "HSPhi"),
        c("Naive T Cells", "Naive"),
        c("Regulatory T Cells", "Regulatory"))

colnames(combined_skin_t_annotations_shorter) <- c("Cluster", "CellType")
combined_skin_t_annotations_shorter <-
  data.frame(combined_skin_t_annotations_shorter)

combined_skin_t <-
  add_annotations(seurat_obj = combined_skin_t,
                  annotations_df = combined_skin_t_annotations_shorter,
                  relabel = FALSE, clusters_col = "annotated_subclusters",
                  annotations_col = "annotated_subclusters_shorter")

colors_annotated_skin_t_shorter <-
  setNames(colors_annotated_skin_t,
           nm = combined_skin_t_annotations_shorter$CellType)
```

```{r}
combined_skin_recombined_annotations_shorter <-
  rbind(c("B Cells", "B Cells"), 
        c("CD4- CD8- T Cells", "CD4- CD8-"), 
        c("CD4+ Effector Memory T Cells", "CD4+ TEMs"), 
        c("CD4+ Tissue-Resident Memory T Cells", "CD4+ TRMs"), 
        c("CD4+ Undefined T Cells", "CD4+ Undefined"), 
        c("CD8+ GZMK+ IFNGhi T Cells", "CD8+ GZMK+ IFNGhi"), 
        c("CD8+ GZMK+ IFNGint T Cells", "CD8+ GZMK+ IFNGint"), 
        c("CD8+ Undefined T Cells", "CD8+ Undefined"), 
        c("Cycling Dendritic Cells", "cDCs"), 
        c("Dividing T Cells", "Dividing"), 
        c("Endothelial Cells", "Endothelial Cells"), 
        c("Fibroblasts", "Fibroblasts"), 
        c("HSPhi T Cells", "HSPhi"), 
        c("Keratinocytes", "Keratinocytes"), 
        c("Langerhans", "Langerhans"), 
        c("Macrophages", "Macrophages"), 
        c("Mast Cells", "Mast Cells"), 
        c("Melanocytes", "Melanocytes"), 
        c("Naive T Cells", "Naive"), 
        c("Natural Killers", "NK Cells"), 
        c("Nerves", "Nerves"), 
        c("Pericytes", "Pericytes"), 
        c("Regulatory T Cells", "Regulatory"))
colnames(combined_skin_recombined_annotations_shorter) <- c("Cluster", "CellType")
combined_skin_recombined_annotations_shorter <-
  data.frame(combined_skin_recombined_annotations_shorter)

combined_skin_recombined <-
  add_annotations(seurat_obj = combined_skin_recombined,
                  annotations_df = combined_skin_recombined_annotations_shorter,
                  relabel = FALSE, clusters_col = "annotated_clusters_expanded",
                  annotations_col = "annotated_clusters_expanded_shorter")
```

## Immune cell types

Immune vs. non-immune:

```{r}
combined_skin_annotations_immune <-
  rbind(c("B Cells", "Immune"),
        c("Cycling Dendritic Cells", "Immune"),
        c("Endothelial Cells", "Non Immune"),
        c("Fibroblasts", "Non Immune"),
        c("Keratinocytes", "Non Immune"),
        c("Langerhans", "Immune"),
        c("Macrophages", "Immune"),
        c("Mast Cells", "Immune"),
        c("Melanocytes", "Non Immune"),
        c("Natural Killers", "Immune"),
        c("Nerves", "Non Immune"),
        c("Pericytes", "Non Immune"),
        c("T Cells", "Immune"))

colnames(combined_skin_annotations_immune) <- c("Cluster", "CellType")
combined_skin_annotations_immune <- data.frame(combined_skin_annotations_immune)

combined_skin <-
  add_annotations(seurat_obj = combined_skin,
                  annotations_df = combined_skin_annotations_immune,
                  relabel = FALSE, clusters_col = "annotated_clusters",
                  annotations_col = "annotated_clusters_immune")
```

Adaptive immune vs. innate immune vs. non-immune:

```{r}
combined_skin_recombined_annotations_immune <-
  rbind(c("B Cells", "Adaptive Immune"),
        c("CD4- CD8- T Cells", "Adaptive Immune"),
        c("CD4+ Effector Memory T Cells", "Adaptive Immune"),
        c("CD4+ Tissue-Resident Memory T Cells", "Adaptive Immune"),
        c("CD4+ Undefined T Cells", "Adaptive Immune"),
        c("CD8+ GZMK+ IFNGhi T Cells", "Adaptive Immune"),
        c("CD8+ GZMK+ IFNGint T Cells", "Adaptive Immune"),
        c("CD8+ Undefined T Cells", "Adaptive Immune"),
        c("Cycling Dendritic Cells", "Innate Immune"),
        c("Dividing T Cells", "Adaptive Immune"),
        c("Endothelial Cells", "Non Immune"),
        c("Fibroblasts", "Non Immune"),
        c("HSPhi T Cells", "Adaptive Immune"),
        c("Keratinocytes", "Non Immune"),
        c("Langerhans", "Innate Immune"),
        c("Macrophages", "Innate Immune"),
        c("Mast Cells", "Innate Immune"),
        c("Melanocytes", "Non Immune"),
        c("Naive T Cells", "Adaptive Immune"),
        c("Natural Killers", "Innate Immune"),
        c("Nerves", "Non Immune"),
        c("Pericytes", "Non Immune"),
        c("Regulatory T Cells", "Adaptive Immune"))

colnames(combined_skin_recombined_annotations_immune) <- c("Cluster", "CellType")
combined_skin_recombined_annotations_immune <-
  data.frame(combined_skin_recombined_annotations_immune)

combined_skin_recombined <-
  add_annotations(seurat_obj = combined_skin_recombined,
                  annotations_df = combined_skin_recombined_annotations_immune,
                  relabel = FALSE, clusters_col = "annotated_clusters_expanded",
                  annotations_col = "annotated_clusters_expanded_immune")
```

Group together CD4+ and CD8+:

```{r}
combined_skin_t_annotations_simpler <-
  rbind(c("CD4- CD8- T Cells", "CD4- CD8- T Cells"),
        c("CD4+ Effector Memory T Cells", "CD4+ T Cells"),
        c("CD4+ Tissue-Resident Memory T Cells", "CD4+ T Cells"),
        c("CD4+ Undefined T Cells", "CD4+ T Cells"),
        c("CD8+ GZMK+ IFNGhi T Cells", "CD8+ T Cells"),
        c("CD8+ GZMK+ IFNGint T Cells", "CD8+ T Cells"),
        c("CD8+ Undefined T Cells", "CD8+ T Cells"),
        c("Dividing T Cells", "Dividing T Cells"),
        c("HSPhi T Cells", "HSPhi T Cells"),
        c("Naive T Cells", "Naive T Cells"),
        c("Regulatory T Cells", "Regulatory T Cells"))

colnames(combined_skin_t_annotations_simpler) <- c("Cluster", "CellType")
combined_skin_t_annotations_simpler <-
  data.frame(combined_skin_t_annotations_simpler)

combined_skin_t <-
  add_annotations(seurat_obj = combined_skin_t,
                  annotations_df = combined_skin_t_annotations_simpler,
                  relabel = FALSE, clusters_col = "annotated_subclusters",
                  annotations_col = "annotated_subclusters_simpler")
```

## Differential expression

```{r}
combined_skin_recombined_annotations_immune <-
  combined_skin_recombined_annotations_immune %>% arrange(desc(CellType))

# do a different order (non immune, then innate, then adaptive)
de_skin_recombined <-
  de_skin_recombined[combined_skin_recombined_annotations_immune$Cluster]

# for the top of the later heatmaps
skin_recombined_types <- combined_skin_recombined_annotations_immune %>%
                           column_to_rownames("Cluster") %>%
                           # nicer for plotting
                           rename("Immune Cell Type" = CellType)
```

## Clean up objects

Add clinical info (we don't use it for any of the manuscript figures, but it might be useful for other researchers to have it be easily accessible):

```{r}
# for consistency
ages <- str_sort(unique(table_clinical$Age), numeric = TRUE)
sexes <- sort(unique(table_clinical$Sex))
sizes <- str_sort(unique(table_clinical$Size), numeric = TRUE)

# blood
combined_blood@meta.data <-
  left_join(combined_blood[[]],
            select(table_clinical, Subject, Age, Sex),
            by = join_by(Subject)) %>%
  mutate(Age = factor(Age, levels = ages),
         Sex = factor(Sex, levels = sexes))
combined_blood_t@meta.data <-
  left_join(combined_blood_t[[]],
            select(table_clinical, Subject, Age, Sex),
            by = join_by(Subject)) %>%
  mutate(Age = factor(Age, levels = ages),
         Sex = factor(Sex, levels = sexes))

# skin
# note that the lesion sizes only make sense for the EM data
combined_skin@meta.data <-
  left_join(combined_skin[[]],
            select(table_clinical, Subject, Age, Sex, Size),
            by = join_by(Subject)) %>%
  mutate(Size = ifelse(SampleType == "EM", Size, NA)) %>%
  mutate(Age = factor(Age, levels = ages),
         Sex = factor(Sex, levels = sexes),
         Size = factor(Size, levels = sizes))
combined_skin_recombined@meta.data <-
  left_join(combined_skin_recombined[[]],
            select(table_clinical, Subject, Age, Sex, Size),
            by = join_by(Subject)) %>%
  mutate(Size = ifelse(SampleType == "EM", Size, NA)) %>%
  mutate(Age = factor(Age, levels = ages),
         Sex = factor(Sex, levels = sexes),
         Size = factor(Size, levels = sizes))
combined_skin_t@meta.data <-
  left_join(combined_skin_t[[]],
            select(table_clinical, Subject, Age, Sex, Size),
            by = join_by(Subject)) %>%
  mutate(Size = ifelse(SampleType == "EM", Size, NA)) %>%
  mutate(Age = factor(Age, levels = ages),
         Sex = factor(Sex, levels = sexes),
         Size = factor(Size, levels = sizes))
```

Fix the rownames:

```{r}
rownames(combined_blood@meta.data) <- combined_blood$Cell_ID_Unique
rownames(combined_blood_t@meta.data) <- combined_blood_t$Cell_ID_Unique
rownames(combined_skin@meta.data) <- combined_skin$Cell_ID_Unique
rownames(combined_skin_recombined@meta.data) <-
  combined_skin_recombined$Cell_ID_Unique
rownames(combined_skin_t@meta.data) <- combined_skin_t$Cell_ID_Unique
```

Rearrange metadata columns to be a little cleaner:

```{r}
# blood
combined_blood@meta.data <-
  combined_blood[[]] %>%
  relocate(cell_id, Cell_ID_Unique, .after = percent.mt) %>%
  relocate(Age, Sex, .after = Disease) %>%
  relocate(Has_TCR, paired_alpha, locus, .after = annotated_clusters) %>%
  relocate(v_call_family_TRA, .after = v_call_gene) %>%
  relocate(j_call_family_TRA, .after = j_call_gene)

combined_blood_t@meta.data <-
  combined_blood_t[[]] %>%
  relocate(cell_id, Cell_ID_Unique, .after = percent.mt) %>%
  relocate(Age, Sex, .after = Disease) %>%
  relocate(Has_TCR, paired_alpha, locus, .after = annotated_clusters) %>%
  relocate(v_call_family_TRA, .after = v_call_gene) %>%
  relocate(j_call_family_TRA, .after = j_call_gene) %>%
  relocate(seurat_subclusters, annotated_subclusters_full,
           annotated_subclusters, .after = last_col())

# skin
combined_skin@meta.data <-
  combined_skin[[]] %>%
  relocate(cell_id, Cell_ID_Unique, .after = percent.mt) %>%
  relocate(Age, Sex, Size, .after = Disease) %>%
  relocate(annotated_clusters_immune, Has_TCR, paired_alpha, locus,
           .after = annotated_clusters) %>%
  relocate(v_call_family_TRA, .after = v_call_gene) %>%
  relocate(j_call_family_TRA, .after = j_call_gene)

combined_skin_recombined@meta.data <-
  combined_skin_recombined[[]] %>%
  relocate(cell_id, Cell_ID_Unique, .after = percent.mt) %>%
  relocate(Age, Sex, Size, .after = Disease) %>%
  relocate(annotated_clusters_expanded, annotated_clusters_expanded_immune,
           Has_TCR, paired_alpha, locus, .after = annotated_clusters) %>%
  relocate(v_call_family_TRA, .after = v_call_gene) %>%
  relocate(j_call_family_TRA, .after = j_call_gene)

combined_skin_t@meta.data <-
  combined_skin_t[[]] %>%
  relocate(cell_id, Cell_ID_Unique, .after = percent.mt) %>%
  relocate(Age, Sex, Size, .after = Disease) %>%
  relocate(Has_TCR, paired_alpha, locus,
           .after = annotated_clusters) %>%
  relocate(v_call_family_TRA, .after = v_call_gene) %>%
  relocate(j_call_family_TRA, .after = j_call_gene) %>%
  relocate(seurat_subclusters, annotated_subclusters_full,
           annotated_subclusters, annotated_subclusters_simpler,
           .after = last_col())
```

We didn't save the original `cell_id`s because we use the `Cell_ID_Unique` columns instead, so let's just drop that column (since it got roped in by using the AIRR data, there are `NA`s for the GEX barcodes that don't have corresponding TCRs):

```{r}
combined_blood$cell_id <- c()
combined_blood_t$cell_id <- c()
combined_skin$cell_id <- c()
combined_skin_recombined$cell_id <- c()
combined_skin_t$cell_id <- c()
```

## Save the objects

For GEO:

```{r}
#| eval = FALSE

# blood
saveRDS(combined_blood,
        file.path(path_objs, "manuscript", "combined_blood.rds"))
saveRDS(combined_blood_t,
        file.path(path_objs, "manuscript", "combined_blood_t.rds"))

# skin
saveRDS(combined_skin,
        file.path(path_objs, "manuscript", "combined_skin.rds"))
saveRDS(combined_skin_recombined,
        file.path(path_objs, "manuscript", "combined_skin_recombined.rds"))
saveRDS(combined_skin_t,
        file.path(path_objs, "manuscript", "combined_skin_t.rds"))
```

------------------------------------------------------------------------

# Tables

## Clinical

192563 refused a control biopsy, the control for 202939 is missing, and we removed the control for 202943 (see 11-sample_removal.Rmd).

```{r table-clinical}
# update the column names for display
table_clinical_display <-
  table_clinical %>%
    rename("Participant Number" = Subject, "Age (years)" = Age,
           "EM Lesion(s)" = EM, "Lesion Size (cm)" = Size)

table_clinical_display <-
  table_clinical_display %>%
  mutate("Paired Control" =
           ifelse(`Participant Number` %in% c("192563", "202939", "202943"),
                  "No", "Yes"))

kable(x = table_clinical_display, align = "c") %>%
  kable_styling("striped") %>%
  scroll_box(height = NULL, width = "100%")

# JCI wants a Word table
# ggpubr::ggtexttable(table_clinical_display, rows = NULL, theme = ttheme("light"))
```

## Counts

Pull info from the QC:

```{r}
# setup the summary table with feature reads
aggr_results_feature_reads <-
  aggr_results[grepl("*_pre_normalization_feature_reads\\>",
                     names(aggr_results))]
names(aggr_results_feature_reads) <- aggr_results[["batches"]] # samples

aggr_results_summary <- as_tibble(aggr_results_feature_reads)
aggr_results_summary$Type <- "pre_normalization_feature_reads"
aggr_results_summary <-
  relocate(aggr_results_summary, Type, .before = "192561SKL")

# add fraction of reads kept
aggr_results_frac_kept <- aggr_results[["frac_reads_kept"]]
names(aggr_results_frac_kept) <- aggr_results[["batches"]]
aggr_results_frac_kept$Type <- "frac_reads_kept"

# add number of cells by library
aggr_results_num_cells <- aggr_results[["num_cells_by_library"]]
names(aggr_results_num_cells) <- aggr_results[["batches"]]
aggr_results_num_cells$Type <- "num_cells_by_library"

aggr_results_summary <- bind_rows(aggr_results_summary, aggr_results_frac_kept,
                                  aggr_results_num_cells)

# transpose the table and add metadata
aggr_results_summary <-
  tibble(aggr_results_summary) %>%
  pivot_longer(cols = c(-Type), names_to = "SampleName") %>%
  pivot_wider(names_from = c(Type)) %>%
  left_join(meta %>% select(SampleName, Subject, SampleType, Dataset),
            by = "SampleName") %>%
  distinct() %>%
  relocate(Dataset, .before = SampleName) %>%
  relocate(Subject, SampleType, .after = SampleName)
```

Pull just the TCR info:

```{r}
metrics_summary_tcr <-
  create_metrics_summary(meta = meta, path_data_specific = path_data_root,
                         data_types = "TCR", samples_list = samples) %>%
  select(SampleName, EstimatedNumberofCells) %>%
  rename(Sample = SampleName, "Cell Count (TCR)" = EstimatedNumberofCells)
```

### Blood

```{r}
table_counts <-
  aggr_results_summary %>%
  # don't include the blood
  filter(SampleType == "Blood") %>%
  # add in metrics summary info
  left_join(metrics_summary %>%
              select(SampleName, ReadsPerCell, MedianGenesperCell),
            by = join_by(SampleName)) %>%
  # make a new column to show the `cellranger aggr` difference
  mutate("Post Aggregation Read Count" =
           round(pre_normalization_feature_reads * frac_reads_kept, digits = 0),
         .after = pre_normalization_feature_reads) %>%
  # round the fraction of reads
  mutate(frac_reads_kept = label_percent(accuracy = 0.1)(frac_reads_kept)) %>%
  # make the names more readable
  rename(Sample = SampleName,
         "Participant Number" = Subject,
         "Sample Type" = SampleType,
         "Pre Aggregation Read Count" = pre_normalization_feature_reads,
         "% Reads Kept" = frac_reads_kept,
         "Cell Count" = num_cells_by_library, "Reads per Cell" = ReadsPerCell,
         "Median Genes per Cell" = MedianGenesperCell)

# add TCR info
table_counts <-
  table_counts %>% left_join(metrics_summary_tcr, by = join_by(Sample))

kable(x = table_counts, align = "c") %>%
  kable_styling("striped") %>%
  scroll_box(height = NULL, width = "100%")

# JCI wants a Word table
# ggpubr::ggtexttable(table_counts, rows = NULL, theme = ttheme("light"))
```

### Skin

```{r}
table_counts <-
  aggr_results_summary %>%
  # don't include the blood
  filter(SampleType != "Blood") %>%
  # add in metrics summary info
  left_join(metrics_summary %>%
              select(SampleName, ReadsPerCell, MedianGenesperCell),
            by = join_by(SampleName)) %>%
  # make a new column to show the `cellranger aggr` difference
  mutate("Post Aggregation Read Count" =
           round(pre_normalization_feature_reads * frac_reads_kept, digits = 0),
         .after = pre_normalization_feature_reads) %>%
  # round the fraction of reads
  mutate(frac_reads_kept = label_percent(accuracy = 0.1)(frac_reads_kept)) %>%
  # make the names more readable
  rename(Sample = SampleName,
         "Participant Number" = Subject,
         "Sample Type" = SampleType,
         "Pre Aggregation Read Count" = pre_normalization_feature_reads,
         "% Reads Kept" = frac_reads_kept,
         "Cell Count" = num_cells_by_library, "Reads per Cell" = ReadsPerCell,
         "Median Genes per Cell" = MedianGenesperCell)

# add TCR info
table_counts <-
  table_counts %>% left_join(metrics_summary_tcr, by = join_by(Sample))

kable(x = table_counts, align = "c") %>%
  kable_styling("striped") %>%
  scroll_box(height = NULL, width = "100%")

# JCI wants a Word table
# ggpubr::ggtexttable(table_counts, rows = NULL, theme = ttheme("light"))
```

------------------------------------------------------------------------

# Figures

## Figure 1 (Skin)

```{r}
# doesn't matter if you use annotated_clusters or annotated_clusters_shorter
plots_figure_1 <- list()

for (sample_type in c("Control", "EM")) {
  combined_skin_sub <- filter(combined_skin[[]], SampleType == sample_type)

  # calculate proportions for each subject, then average them
  props_skin_sub <-
    combined_skin_sub %>%
    group_by(Subject, annotated_clusters, annotated_clusters_immune) %>%
    summarize(count = n()) %>%
    ungroup() %>%
    group_by(Subject) %>%
    mutate(proportion = count / sum(count)) %>%
    ungroup() %>%
    group_by(annotated_clusters, annotated_clusters_immune) %>%
    summarize(mean_proportion = mean(proportion)) %>%
    ungroup() %>%
    mutate(SampleType = sample_type)

  # reorder clusters by immune type and mean_proportion within each SampleType
  props_skin_sub_order <-
    props_skin_sub %>%
    arrange(SampleType, desc(annotated_clusters_immune), mean_proportion) %>%
    pull(annotated_clusters)
  props_skin_sub <-
    props_skin_sub %>%
      mutate(annotated_clusters =
               factor(annotated_clusters, levels = props_skin_sub_order))
  
  # calculate where to split the immune types
  props_skin_sub_immune <- props_skin_sub %>%
                             group_by(annotated_clusters_immune) %>%
                             summarize(n = sum(mean_proportion)) %>%
                             arrange(annotated_clusters_immune) %>% 
                             pull(n)
  props_skin_sub_immune <- props_skin_sub_immune[1] # just need one
  bar_height <- 1.55
  
  plots_figure_1[[sample_type]] <-
    ggplot(props_skin_sub,
           aes(x = SampleType, y = mean_proportion, fill = annotated_clusters)) +
      geom_bar(stat = "identity", color = "black", linewidth = 0.2) +
      geom_segment(aes(color = "Immune"), x = bar_height, xend = bar_height,
                   y = 0, yend = props_skin_sub_immune, linewidth = 4) +
      geom_segment(aes(color = "Non Immune"), x = bar_height, xend = bar_height,
                   y = props_skin_sub_immune, yend = 1, linewidth = 4) +
      labs(x = sample_type, y = "Cell Type Percentage", fill = "Cell Type",
           color = "Category") +
      coord_flip() +
      scale_y_continuous(labels = label_percent(scale = 100),
                         expand = expansion(mult = 0.01)) +
      scale_fill_manual(breaks = names(colors_annotated_skin),
                        values = colors_annotated_skin, guide = "none") +
      scale_color_manual(values = colors_immune) +
      geom_label(aes(label =
                       ifelse(mean_proportion > 0.03,
                              label_percent(accuracy = 1)(mean_proportion), NA),
                     group = annotated_clusters),
                 color = "black", fill = "white", size = 4.5,
                 position = position_stack(vjust = 0.5)) +
      theme_bw + labels_standard + labels_manuscript_prop
}
```

```{r figure-1-umap-freq-skin}
#| fig.dim = c(17, 18)

p1 <- plot_umap(seurat_obj = combined_skin, tissue_type = "Skin",
                use_hues = TRUE, plot_by = "cluster_all", label_size = 3.5) +
        guides(color = guide_legend(override.aes = list(size = 3), ncol = 2)) +
        labels_manuscript_text

p2 <- plot_umap(seurat_obj = combined_skin, tissue_type = "Skin",
                clrs_specific = colors_annotated_skin_shorter,
                plot_by = "cluster_all", label_size = 3.5, annotated = TRUE,
                annotations_col = "annotated_clusters_shorter") +
        labels_manuscript_text

# highlight increasing immune cells
p3 <- plot_subsets(seurat_obj = combined_skin,
                   tissue_type = "Skin",
                   clrs_specific = colors_sample_types_manuscript,
                   facet_by = "annotated_clusters_shorter",
                   # don't need Dataset
                   meta_group_by = c("Subject", "SampleType"),
                   highlights = c("B Cells", "Macrophages", "NK Cells",
                                  "T Cells"),
                   signif_caption = FALSE, signif_size = 6) +
        labs(title = NULL, y = NULL) +
        labels_manuscript_text +
        theme(strip.text.x = element_text(size = 13))

# combine the plots
plots_figure_1_bar <-
  (plots_figure_1$Control / plots_figure_1$EM) + plot_layout(guides = "collect")

(((p1 | p2) / free(plots_figure_1_bar) / free(p3)) +
    plot_anno + plot_layout(heights = c(3, 2, 3.5))) &
    labels_manuscript_text_big
```

Save for SDV spreadsheet:

```{r}
sdv$Figure_1C <- plots_figure_1$Control$data
sdv$Figure_1D <- plots_figure_1$EM$data
sdv$Figure_1E <- p3$data
```

Just for the significance values:

```{r figure-1-freq}
#| fig.dim = c(15, 7),
#| eval = FALSE

plot_subsets(seurat_obj = combined_skin,
             tissue_type = "Skin",
             clrs_specific = colors_sample_types_manuscript,
             facet_by = "annotated_clusters",
             highlights = c("B Cells", "Macrophages", "Natural Killers",
                            "T Cells"),
             signif_stars = FALSE) +
    labs(title = NULL, y = NULL)
```

## Figure 2 (Skin T cells)

*Not ordered by immune type*

```{r}
plots_figure_2 <- list()

for (sample_type in c("Control", "EM")) {
  combined_skin_t_sub <- filter(combined_skin_t[[]], SampleType == sample_type)
  
  props_skin_t_sub <-
  (sweep(table(combined_skin_t_sub$Subject,
               combined_skin_t_sub$annotated_subclusters), 1,
       table(combined_skin_t_sub$Subject), FUN = "/") * 100) %>%
    as.data.frame() %>%
    group_by(Var2) %>%
    mutate(Freq = sum(Freq)) %>%
    select(-Var1) %>% distinct() %>% ungroup() %>%
    mutate(mean_proportion = Freq/(sum(Freq)), SampleType = "Control") %>%
    rename(annotated_subclusters = Var2) %>% select(-Freq)
  
  # reorder clusters by immune type and mean_proportion within each SampleType
  props_skin_t_sub_order <-
    props_skin_t_sub %>%
    arrange(SampleType, mean_proportion) %>%
    pull(annotated_subclusters)
  props_skin_t_sub <-
    props_skin_t_sub %>%
      mutate(annotated_subclusters =
               factor(annotated_subclusters, levels = props_skin_t_sub_order))
  
  plots_figure_2[[sample_type]] <-
    ggplot(props_skin_t_sub,
           aes(x = SampleType, y = mean_proportion, fill = annotated_subclusters)) +
      geom_bar(stat = "identity", color = "black", linewidth = 0.2) +
      labs(x = sample_type,
           y = "Cell Type Percentage", fill = "Cell Type") +
      coord_flip() +
      scale_y_continuous(labels = label_percent(scale = 100),
                         expand = expansion(mult = 0.01)) +
      scale_fill_manual(breaks = names(colors_annotated_skin_t),
                        values = colors_annotated_skin_t) +
      geom_label(aes(label =
                       ifelse(mean_proportion > 0.03,
                              label_percent(accuracy = 1)(mean_proportion), NA),
                     group = annotated_subclusters),
                 color = "black", fill = "white", size = 4.5,
                 position = position_stack(vjust = 0.5)) +
      theme_bw + labels_standard + labels_manuscript_prop +
      theme(legend.position = "none")
}
```

Subsets as part of the full skin:

```{r figure-2-umap-freq-skin-t}
#| fig.dim = c(17, 18)

p1 <- plot_umap(seurat_obj = combined_skin_t, tissue_type = "Skin T Cells",
                use_hues = TRUE, plot_by = "cluster_all",
                clusters_col = "seurat_subclusters", label_size = 3.5) +
        labels_manuscript_text

p2 <- plot_umap(seurat_obj = combined_skin_t, tissue_type = "Skin T Cells",
                clrs_specific = colors_annotated_skin_t_shorter,
                plot_by = "cluster_all", annotated = TRUE,
                annotations_col = "annotated_subclusters_shorter",
                label_size = 3.5) +
        labels_manuscript_text

# highlight increasing immune cells
t_cell_types <- levels(combined_skin_t$annotated_subclusters_shorter)
# has to be updated manually unfortunately
increasing_cells <- t_cell_types[!grepl("HSPhi|Undefined", t_cell_types)]
p3 <- plot_subsets(seurat_obj = combined_skin_recombined,
                   tissue_type = "Skin T Cells",
                   clrs_specific = colors_sample_types_manuscript,
                   facet_by = "annotated_clusters_expanded_shorter",
                   # don't need Dataset
                   meta_group_by = c("Subject", "SampleType"),
                   ncol = 6, filter_to = t_cell_types,
                   highlights = increasing_cells, signif_caption = FALSE,
                   signif_size = 6) +
        labs(title = NULL, y = NULL) +
        labels_manuscript_text +
        # to make the long names fit
        theme(strip.text.x = element_text(size = 13))

# combine the plots
plots_figure_2_bar <-
  (plots_figure_2$Control / plots_figure_2$EM)

((p1 | p2) / free(plots_figure_2_bar) / free(p3)) +
  plot_anno + plot_layout(heights = c(3, 2, 3.5)) &
  labels_manuscript_text_big
```

Save for SDV spreadsheet:

```{r}
sdv$Figure_2C <- plots_figure_2$Control$data
sdv$Figure_2D <- plots_figure_2$EM$data
sdv$Figure_2E <- p3$data
```

Just for the significance values:

```{r figure-2-freq}
#| fig.dim = c(15, 7),
#| eval = FALSE

plot_subsets(seurat_obj = combined_skin_recombined,
             tissue_type = "Skin T Cells",
             clrs_specific = colors_sample_types_manuscript,
             facet_by = "annotated_clusters_expanded", ncol = 6,
             filter_to = levels(combined_skin_t$annotated_subclusters),
             # has to be updated manually unfortunately
             highlights = c("CD4- CD8- T Cells",
                            "CD4+ Effector Memory T Cells",
                            "CD4+ Tissue-Resident Memory T Cells",
                            "CD8+ GZMK+ IFNGhi T Cells",
                            "CD8+ GZMK+ IFNGint T Cells",
                            "Dividing T Cells",
                            "HSPhi T Cells",
                            "Naive T Cells",
                            "Regulatory T Cells"),
             signif_stars = FALSE) +
  labs(title = NULL, y = NULL) +
  theme(strip.text.x = element_text(size = 9))
```

## Figure 3 (IFNGhi vs IFNGint)

```{r figure-3-waterfall}
#| fig.dim = c(12, 5)

# for plotting purposes
de_genes_skin_t_em_ifng_waterfall <-
  de_genes_skin_t_em_ifng %>%
  arrange(desc(avg_logFC)) %>%
  mutate(comparison = "CD8+ GZMK+ IFNGhi vs. IFNGint T Cells",
         gene = factor(gene, levels = unique(gene)),
         annotated_subclusters =
           ifelse(avg_logFC > 0,
                  "CD8+ GZMK+ IFNGhi T Cells", "CD8+ GZMK+ IFNGint T Cells"))

# save for SDV spreadsheet
sdv$Figure_3 <- de_genes_skin_t_em_ifng_waterfall

# just color by cell type
ggplot(de_genes_skin_t_em_ifng_waterfall) +
        aes(x = gene, y = avg_logFC, fill = annotated_subclusters) +
  geom_col(linewidth = 0.2, color = "black") +
  scale_fill_manual(name = "Cell Type with Higher Expression",
                    values = colors_annotated_skin_t) +
  labs(x = "Differentially Expressed Genes", y = "log2FC") +
  theme_bw + labels_standard + labels_rotate_x +
  labels_manuscript + labels_manuscript_text + theme(legend.position = "top")
```

## Figure 4 (Clonal overlaps)

```{r}
plots_figure_4 <- list()

# only show a few cell types in the main figure
specific_cell_types_skin_simpler <- c("CD4+ T Cells", "CD8+ T Cells")
specific_cell_types_skin_gzmk <-
  c("CD8+ GZMK+ IFNGhi T Cells", "CD8+ GZMK+ IFNGint T Cells")
specific_cell_types_skin <-
  c(specific_cell_types_skin_gzmk, "Regulatory T Cells")
```

```{r}
# pull clones and remove NAs
clones_blood <-
  combined_blood_t[[]] %>%
    filter(!is.na(clone_id_unique)) %>%
    pull(clone_id_unique)
clones_control <-
  combined_skin_t[[]] %>%
    filter(!is.na(clone_id_unique), SampleType == "Control") %>%
    pull(clone_id_unique)
clones_em <-
  combined_skin_t[[]] %>%
    filter(!is.na(clone_id_unique), SampleType == "EM") %>%
    pull(clone_id_unique)

list_venn <-
  list("Blood" = clones_blood, "Control" = clones_control, "EM" = clones_em)
list_venn_total <- Venn(list_venn)

plots_figure_4[["venn_total"]] <-
  ggVennDiagram(x = list_venn,
                set_color = colors_sample_types_manuscript, set_size = 0,
                label = "count", label_alpha = 0, label_size = 5,
                edge_size = 1.5) +
    geom_text(data = venn_setlabel(process_data(Venn(list_venn))),
              aes(X, Y, label = name, size = 5), show.legend = FALSE) +
    labs(title = "All Clones in the Blood and Skin T Cells",
         fill = "Clone Count") +
    scale_fill_gradient2(low = "white", high = "grey50") +
    # don't cut off "Control"
    scale_x_continuous(expand = expansion(mult = 0.2)) +
    labels_standard_venn +
    theme(plot.title = element_text(size = 14))
```

```{r figure-4-clonal-venn-skin-t-simpler}
plots_venn_skin_t_simpler <- list()
list_venn_skin_t_simpler <- list()

for (cell_type in specific_cell_types_skin_simpler) {
  combined_skin_t_sub <-
    subset(combined_skin_t, subset = annotated_subclusters_simpler == cell_type)
  
  # pull clones and remove NAs
  clones_control <-
    combined_skin_t_sub[[]] %>%
      filter(!is.na(clone_id_unique), SampleType == "Control") %>%
      pull(clone_id_unique)
  clones_em <-
    combined_skin_t_sub[[]] %>%
      filter(!is.na(clone_id_unique), SampleType == "EM") %>%
      pull(clone_id_unique)
  # we only want the overlapping clones (because it makes no sense to show blood
  # clones associated with a skin cell type)
  clones_blood <-
    combined_blood_t[[]] %>%
      filter(clone_id_unique %in% c(clones_control, clones_em)) %>%
      pull(clone_id_unique)
  
  list_venn <-
    list("Blood" = clones_blood, "Control" = clones_control, "EM" = clones_em)
  
    # for ease of counting for the results
  list_venn_skin_t_simpler[[cell_type]] <- Venn(list_venn)

  plots_venn_skin_t_simpler[[cell_type]] <-
    ggVennDiagram(x = list_venn,
                  set_color = colors_sample_types_manuscript, set_size = 0,
                  label = "count", label_alpha = 0, label_size = 5,
                  edge_size = 1.5) +
      geom_text(data = venn_setlabel(process_data(Venn(list_venn))),
                aes(X, Y, label = name, size = 5), show.legend = FALSE) +
      labs(title = paste("Skin", cell_type, "Clones"),
           fill = "Clone Count") +
      scale_fill_gradient2(low = "white", high = "grey50") +
      # don't cut off "Control"
      scale_x_continuous(expand = expansion(mult = 0.2)) +
      labels_standard_venn +
      theme(plot.title = element_text(size = 14))
}

plots_figure_4[["venn_skin_simpler"]] <-
  plots_venn_skin_t_simpler[specific_cell_types_skin_simpler]
```

```{r figure-4-clonal-venn-skin-t}
plots_venn_skin_t <- list()
list_venn_skin_t <- list()

for (cell_type in sort(unique(combined_skin_t$annotated_subclusters))) {
  combined_skin_t_sub <-
    subset(combined_skin_t, subset = annotated_subclusters == cell_type)
  
  # pull clones and remove NAs
  clones_control <-
    combined_skin_t_sub[[]] %>%
      filter(!is.na(clone_id_unique), SampleType == "Control") %>%
      pull(clone_id_unique)
  clones_em <-
    combined_skin_t_sub[[]] %>%
      filter(!is.na(clone_id_unique), SampleType == "EM") %>%
      pull(clone_id_unique)
  # we only want the overlapping clones (because it makes no sense to show blood
  # clones associated with a skin cell type)
  clones_blood <-
    combined_blood_t[[]] %>%
      filter(clone_id_unique %in% c(clones_control, clones_em)) %>%
      pull(clone_id_unique)
  
  list_venn <-
    list("Blood" = clones_blood, "Control" = clones_control, "EM" = clones_em)
  
    # for ease of counting for the results
  list_venn_skin_t[[cell_type]] <- Venn(list_venn)

  plots_venn_skin_t[[cell_type]] <-
    ggVennDiagram(x = list_venn,
                  set_color = colors_sample_types_manuscript, set_size = 0,
                  label = "count", label_alpha = 0, label_size = 5,
                  edge_size = 1.5) +
      geom_text(data = venn_setlabel(process_data(Venn(list_venn))),
                aes(X, Y, label = name, size = 5), show.legend = FALSE) +
      labs(title = paste("Skin", cell_type, "Clones"),
           fill = "Clone Count") +
      scale_fill_gradient2(low = "white", high = "grey50") +
      # don't cut off "Control"
      scale_x_continuous(expand = expansion(mult = 0.2)) +
      labels_standard_venn +
      theme(plot.title = element_text(size = 14))
}

plots_figure_4[["venn_skin"]] <- plots_venn_skin_t[specific_cell_types_skin]
```

```{r figure-4-clonal-venn}
#| fig.dim = c(14, 7)

# Venn diagrams
wrap_plots(list_flatten(list(plots_figure_4$venn_total,
                             plots_figure_4$venn_skin_simpler,
                             plots_figure_4$venn_skin)), nrow = 2) +
  plot_anno & labels_manuscript_venn
```

## Figure 5 (Chemokines)

```{r figure-5a-chemokine-heatmap}
#| fig.dim = c(9, 7.5)

genes_list_ligands <- c("CCL4", "CCL5", "CXCL9", "CXCL10", "CXCL11", "CXCL13",
                        "XCL1", "XCL2")
genes_list_receptors <- c("CCR1", "CCR3", "CCR5", "CXCR5")

# all genes together
genes_list <- c(genes_list_ligands, genes_list_receptors)

# and split the rows by group (so much nicer than patchwork)
chemokine_groups <-
    data.frame("Type" = c(rep("Ligands", length(genes_list_ligands)),
                          rep("Receptors", length(genes_list_receptors))),
               "Genes" = genes_list)

heatmap_DESeq2(DESeq2_out = de_skin_recombined, gene_list = genes_list,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types, clrs_anno_top = colors_immune,
               split_row = chemokine_groups,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

```{r figure-5bc-chemokine-circos}
#| eval = FALSE

pathways.show <- c("CCL", "CXCL")

# reset graphics device cleanly
try(graphics.off(), silent = TRUE)

# plot grid + error-guard each panel
op <- par(no.readonly = TRUE); on.exit(par(op), add = TRUE)
par(mfcol = c(1, length(pathways.show)), # one row with two columns
    xpd = TRUE, mar = c(1.5, 1.5, 2, 2)) # bottom, left, top, right

# EM only
plot_list <- list()
for (pathway in pathways.show) {
  # there's no option to remove the labels, so we'll make them tiny and white
  # and get rid of them manually in Adobe Illustrator
  netVisual_aggregate(object = cellchat_list$EM, signaling = pathway,
                      signaling.name = paste(pathway, "in EM"),
                      color.use = colors_cellchat,
                      remove.isolate = TRUE,
                      layout = "circle", pt.title = 14,
                      vertex.label.cex = 0.1, alpha.edge = 0.9,
                      vertex.label.color = "white",
                      edge.curved = FALSE, arrow.size = 0.4, text.y = 1.3)

  # add the plot to the list
  plot_list[[pathway]] <- recordPlot()
}

# save the plot manually (since base R doesn't output through R Markdown)
png(filename = here::here("primary_analysis", "figures", "manuscripts",
                          "manuscript",
                          paste0("manuscript-",
                                 "figure-5bc-chemokine-circos-1.png")),
    width = 18, height = 9, units = "in", res = 600)
for (p in plot_list) {
  plot.new()
  print(p)
}
dev.off()
```

```{r}
include_graphics(path =
                   here::here("primary_analysis", "figures", "manuscripts",
                              "manuscript",
                              paste0("manuscript-",
                                     "figure-5bc-chemokine-circos-1.png")))
```

Save for SDV spreadsheet:

```{r}
# pulls from the first lines of heatmap_DESeq2()
sdv$Figure_5A <-
  enframe(de_skin_recombined) %>%
  unnest(value) %>%
  rename(annotated_clusters_expanded = name) %>%
  dplyr::filter(gene %in% genes_list) %>%
  # need the genes to be alphanumerically sorted
  mutate(gene = factor(gene,
                       levels = str_sort(unique(gene), numeric = TRUE))) %>%
  arrange(annotated_clusters_expanded, gene)

# pulls from netVisual_aggregate()
get_cellchat_strength <- function(pathway) {
  object <- cellchat_list$EM
  signaling <- pathway
  pairLR <- searchPair(signaling = signaling, pairLR.use = object@LR$LRsig, 
                       key = "pathway_name", matching.exact = TRUE,
                       pair.only = TRUE)
  net <- object@net
  pairLR.use.name <- dimnames(net$prob)[[3]]
  pairLR.name <- intersect(rownames(pairLR), pairLR.use.name)
  prob <- net$prob
  prob[net$pval > 0.05] <- 0
  pairLR.name.use <- pairLR.name[apply(prob[, , pairLR.name], 3, sum) != 0]
  prob <- prob[, , pairLR.name.use]
  prob.sum <- as.data.frame(apply(prob, c(1, 2), sum))
  
  prob.sum %>%
    rownames_to_column(var = "Senders") %>%
    pivot_longer(cols = names(prob.sum), names_to = "Receivers",
                 values_to = "Strength") %>%
    filter(Strength != 0)
}

sdv$Figure_5B <- get_cellchat_strength(pathway = "CCL")
sdv$Figure_5C <- get_cellchat_strength(pathway = "CXCL")
```

## Figure 6 (GSEA)

```{r}
sig_gsea_results <- c()

for (i in names(de_skin_recombined)) {
  degs <- data.frame(gene = de_skin_recombined[[i]]$gene,
                     stat = de_skin_recombined[[i]]$stat) %>%
            arrange(desc(stat))
  gene_list <- degs$stat
  names(gene_list) <- degs$gene
  
  msig_gsea <-
    clusterProfiler::GSEA(gene_list, 
                          scoreType = "pos", # std, neg (no results)
                          pvalueCutoff = 0.01, 
                          pAdjustMethod = "bonferroni",
                          TERM2GENE = hallmark, verbose = FALSE)
  
  if (dim(msig_gsea)[1] > 0) {
    sig_gsea_results <-
      bind_rows(sig_gsea_results,
                msig_gsea@result %>% mutate(Cells = i) %>% remove_rownames())
  }
}

# easier to read pathways
sig_gsea_simpler <- sig_gsea_results %>%
                      select(-ID) %>% 
                      # keep the original descriptions
                      rename(Pathways = Description, p_adjust = p.adjust) %>%
                      mutate(Description = Pathways,
                             Pathways =
                                 str_remove(Pathways, "HALLMARK_") %>%
                                 str_replace_all("_", " "))
sig_gsea_simpler <- left_join(sig_gsea_simpler,
                              combined_skin_recombined_annotations_immune %>%
                                rename(Cells = Cluster),
                              by = join_by(Cells))

# let's get rid of the all caps actually
# we want to keep a few words in upper case
caps_words <- c("AKT", "DNA", "E2F", "G2M", "IL2", "IL6", "JAK", "KRAS", "MYC",
                "P53", "PI3K", "STAT3", "STAT5", "TGF", "UV")
caps_words <- setNames(caps_words, str_to_title(caps_words))
# there are a few special cases
caps_words <- c(caps_words, "Mtor" = "mTOR", "Mtorc1" = "mTORC1",
                "Nfkb" = "NFkB", "Tnfa" = "TNFa", "V1" = "v1", "V2" = "v2",
                "Via" = "via")

sig_gsea_simpler <- sig_gsea_simpler %>%
                      mutate(Pathways = str_to_title(Pathways),
                             Pathways = str_replace_all(Pathways, caps_words))

# save for SDV spreadsheet
sdv$Figure_6 <- sig_gsea_simpler %>%
                  arrange(Cells, Pathways) %>%
                  relocate(Cells, CellType, Description, .before = Pathways) %>%
                  rename(annotated_clusters_expanded = Cells,
                         annotated_clusters_immune = CellType)
```

```{r figure-6-gsea-pathways-skin-point}
#| fig.dim = c(18, 11)

# use purple to not conflict with other heatmap blue
ggplot(sig_gsea_simpler, aes(x = Cells, y = Pathways, color = p_adjust)) +
  geom_point(aes(fill = p_adjust), color = "black", size = 5, pch = 21) +
  labs(x = NULL, y = NULL) +
  scale_x_discrete(limits = rev) + # alphabetical order
  scale_fill_gradientn(colors = c("#aa8cfa", "white")) +
  coord_flip() +
  facet_grid(rows = vars(CellType), scales = "free_y", space = "free_y") +
  theme_bw + labels_standard + labels_rotate_x +
  labels_manuscript_text_big + theme(panel.grid.major.x = element_line(),
                                     strip.text.y = element_text(size = 16))
```

## Figure 7 (Interferon response and production)

```{r figure-7-heatmap}
#| fig.dim = c(9, 12.5)

# we only want the protein-coding genes, so let's list them manually by deleting
# from the commented out code's output
# genes_list <- str_sort(grep("^IFN", rownames(combined_skin_recombined), 
#                             value = TRUE), numeric = TRUE)
genes_list_ifn <- c(# Type I
                    "IFNA1", "IFNA2", "IFNA4", "IFNA5", "IFNA6", "IFNA7", 
                    "IFNA8", "IFNA10", "IFNA13", "IFNA14", "IFNA16", "IFNA17",
                    "IFNA21", "IFNB1", "IFNE", "IFNK", "IFNW1",
                    # Type II
                    "IFNG",
                    # Type III
                    "IFNL1", "IFNL2", "IFNL3", "IFNL4")

# let's include the ISG genes too
genes_list_irg <- c("IFITM1", "IFITM2", "IRF7", "ISG15", "ISG20", "ISG20L2",
                    "SOCS3", "STAT1", "STAT2", "STAT3")

# and the interleukins
genes_list_il <- c("IL6", "IL10", "IL12A", "IL12B", "IL15", "IL16", "IL18",
                   "IL32", "TGFB1")

# no JAK or TYK
genes_list_signaling <- c("IRF1", "NFKB1", "NFKB2", "TNF")

# all genes together
genes_list <-
  c(genes_list_ifn, genes_list_irg, genes_list_il, genes_list_signaling)

# and split the rows by group (so much nicer than patchwork)
ifn_groups <-
  data.frame("Type" = c(rep("Interferon", length(genes_list_ifn)),
                        rep("Interferon-Regulated", length(genes_list_irg)),
                        rep("Interleukin and TGFB", length(genes_list_il)),
                        rep("TNFA Signaling", length(genes_list_signaling))),
             "Genes" = genes_list)

heatmap_DESeq2(DESeq2_out = de_skin_recombined, gene_list = genes_list,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types, clrs_anno_top = colors_immune,
               split_row = ifn_groups,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

Save for SDV spreadsheet:

```{r}
# pulls from the first lines of heatmap_DESeq2() 
sdv$Figure_7 <-
  enframe(de_skin_recombined) %>%
  unnest(value) %>%
  rename(annotated_clusters_expanded = name) %>%
  dplyr::filter(gene %in% genes_list) %>%
  # need the genes to be alphanumerically sorted
  mutate(gene = factor(gene,
                       levels = str_sort(unique(gene), numeric = TRUE))) %>%
  arrange(annotated_clusters_expanded, gene)
```

------------------------------------------------------------------------

# Supplemental

For the dot plots, we could combine the cluster and cell type labels into one.

## Setup

Custom lists of markers:

```{r}
markers_blood <-
  list("B Cells" = c("CD19", "CD27", "IGHD", "IGHM", "MS4A1"),
       "Dendritic Cells" = c("CD1C", "CLEC4C", "CLEC10A", "FCER1A"),
       "Monocytes" = c("CD14", "FCGR3A", "FCN1", "LYZ", "S100A9"),
       "Natural Killers" = c("FCGR3A", "GNLY", "GZMA", "KLRB1", "NCAM1", "NKG7"),
       "Plasma Cells" = c("IGHG1", "IGKC", "MZB1", "XBP1"),
       "Platelets" = c("PF4", "PPBP", "TUBB1"),
       "T Cells" = c("CCR7", "CD3D", "CD4", "CD8A", "CD8B", "IL7R"))
markers_blood <- data.frame(Cell_Type_Full = rep(names(markers_blood),
                                                 sapply(markers_blood, length)),
                            features.plot = unlist(markers_blood)) %>%
                  remove_rownames()

markers_blood_t <-
    list("CD4 T Cells" = c("CCR7", "CD4"),
         "CD8 T Cells" = c("CD8A", "CD8B"),
         "GZMK Cells" = c("GZMB", "GZMK"),
         "MAIT Cells" = c("KLRB1", "SLC4A10", "TRAV1-2"),
         "Gamma Delta T Cells" = c("TRDC", "TRDV2", "TRGC1", "TRGV9", "TRGV10"),
         "Proliferating Cells" = c("CDK1", "MKI67", "PCNA", "TOP2A"))
markers_blood_t <-
    data.frame(Cell_Type_Full = rep(names(markers_blood_t),
                                    sapply(markers_blood_t, length)),
               features.plot = unlist(markers_blood_t)) %>%
    remove_rownames()

markers_skin <-
  list("B Cells" = c("CD19", "CD27", "IGHD", "IGHM", "MS4A1"),
       "Cycling Cells" = c("MKI67", "TOP2A"),
       "Dendritic Cells" = c("CLEC9A", "ITGAE", "ITGAX", "XCR1"),
       "Endothelial Cells" = c("CD93", "PECAM1", "SELE", "VWF"),
       "Fibroblasts" = c("COL1A1", "COL6A2", "DCN", "LUM", "PDGFRA"),
       "Keratinocytes" = c("KRT1", "KRT5", "KRT10", "KRT14"),
       "Langerhans" = c("CD1A", "CD207"),
       "Macrophages" = c("CD14", "CD163", "CD68", "CTSS", "ITGAM", "MS4A7"),
       "Mast Cells" = c("CPA3", "IL1RL1", "TPSAB1"),
       "Melanocytes" = c("DCT", "MLANA", "PMEL", "TYRP1"),
       "Natural Killers" = c("FCGR3A", "GNLY", "GZMA", "KLRB1", "NCAM1", "NKG7"),
       "Nerves" = c("CDH19", "MPZ", "S100B", "SCN7A"),
       "Pericytes" = c("CCL2", "IL6", "RGS5"),
       "T Cells" = c("CD3D", "CD4", "CD8A", "CD8B", "IL7R"))
markers_skin <-
  data.frame(Cell_Type_Full = rep(names(markers_skin),
                                  sapply(markers_skin, length)),
             features.plot = unlist(markers_skin)) %>%
  remove_rownames()

# PD-1/PDCD1 is duplicated since it's a general exhaustion marker
markers_skin_t <-
    list("CD4 T Cells" = c("CD3D", "CD4", "CD40LG"),
         "CD8 GZMK IFNG T Cells" = c("CD8A", "CD8B", "GZMB", "GZMK", "IFNG",
                                     "PRF1"),
         "Dividing Cells" = c("CDK1", "MKI67", "PCNA", "TOP2A"),
         "Effector Memory T Cells" = c("CTLA4", "EOMES", "FOS", "GZMH", "IL7R",
                                       "KLRB1"),
         "Heat Shock Protein Cells" = c("HSPA1A", "HSPA6", "HSPB1"),
         "Naive T Cells" = c("CCR7", "IL2RG", "TCF7", "SELL"),
         "Regulatory T Cells" = c("FOXP3", "IL2RA", "TIGIT"),
         "Tissue-Resident T Cells" = c("CCL5", "CD69", "CXCR6", "GZMA", "ITGA4",
                                       "ITGAE", "PDCD1", "RGS1", "TBX21"))
markers_skin_t <-
    data.frame(Cell_Type_Full = rep(names(markers_skin_t),
                                    sapply(markers_skin_t, length)),
               features.plot = unlist(markers_skin_t)) %>%
    remove_rownames()
```

## Supplemental Figure 1 (Skin)

```{r figure-supplemental-1-skin-dot}
#| fig.dim = c(16, 15)

p1 <- DotPlot(combined_skin,
              features = unique(markers_skin$features.plot), 
              dot.scale = 3, group.by = "seurat_clusters") %>%
        dot_color_scale() %>%
        add_info_bar(method = "join", info_type = "Cell_Type_Full",
                     info = markers_skin,
                     text_size = 12, label = FALSE, angle = 90) %>%
        plot_dot_side(seurat_obj = combined_skin,
                      row_identity = "seurat_clusters",
                      facet_col = "Cell_Type_Full",
                      info_to_add = c("cluster_size", "percent_TCR")) +
        labels_manuscript & labels_manuscript_text & labels_manuscript_dot +
        theme(axis.text.x = element_blank())

# a bit of the cDCs and NKs show up in the nerve section
p2 <- DotPlot(combined_skin,
              features = unique(markers_skin$features.plot),
              cols = colors_dot, dot.scale = 3) %>%
        dot_color_scale() %>%
        add_info_bar(method = "join", info_type = "Cell_Type_Full",
                     info = markers_skin,
                     text_size = 12, label = FALSE, angle = 90) %>%
        plot_dot_side(seurat_obj = combined_skin,
                      row_identity = "annotated_clusters",
                      facet_col = "Cell_Type_Full",
                      info_to_add = c("cluster_size", "percent_TCR")) +
        labels_manuscript & labels_manuscript_text & labels_manuscript_dot +
        theme(axis.text.x = element_text(angle = 90),
              strip.text.x = element_blank())

# combine the plots (and make the legend order consistent)
(p1 / p2) + plot_anno + plot_layout(heights = c(3, 1)) &
  # we have to redo the titles since they change when the legends are altered
  guides(size = guide_legend(title = "Percent Expressed", order = 1),
         color = guide_colorbar(title = "Average Expression", order = 2)) &
  labs(title = NULL, subtitle = NULL, x = NULL, y = NULL)

# save for SDV spreadsheet
sdv$Supplemental_Figure_1A <-
  p1$data %>%
  arrange(id) %>%
  relocate(id, Cell_Type_Full, features.plot, pct.exp, .before = avg.exp) %>%
  rename(seurat_clusters = id)
sdv$Supplemental_Figure_1B <-
  p2$data %>%
  arrange(id) %>%
  relocate(id, Cell_Type_Full, features.plot, pct.exp, .before = avg.exp) %>%
  rename(annotated_clusters = id)
```

## Supplemental Figure 2 (Skin T cells)

```{r figure-supplemental-2-skin-t-dot}
#| fig.dim = c(16, 12)

p1 <- DotPlot(combined_skin_t,
              features = unique(markers_skin_t$features.plot),
              dot.scale = 3, group.by = "seurat_subclusters") %>%
        dot_color_scale() %>%
        add_info_bar(method = "join", info_type = "Cell_Type_Full",
                     info = markers_skin_t, sort = FALSE,
                     text_size = 12, label = FALSE, angle = 90) %>%
        plot_dot_side(seurat_obj = combined_skin_t,
                      row_identity = "seurat_subclusters",
                      facet_col = "Cell_Type_Full",
                      info_to_add = c("cluster_size", "percent_TCR")) +
        labels_manuscript & labels_manuscript_text & labels_manuscript_dot +
        theme(axis.text.x = element_blank())

p2 <- DotPlot(combined_skin_t,
              features = unique(markers_skin_t$features.plot),
              dot.scale = 3) %>%
        dot_color_scale() %>%
        add_info_bar(method = "join", info_type = "Cell_Type_Full",
                     info = markers_skin_t, sort = FALSE,
                     text_size = 12, label = FALSE, angle = 90) %>%
        plot_dot_side(seurat_obj = combined_skin_t,
                      row_identity = "annotated_subclusters",
                      facet_col = "Cell_Type_Full",
                      info_to_add = c("cluster_size", "percent_TCR")) +
        labels_manuscript & labels_manuscript_text & labels_manuscript_dot +
        theme(axis.text.x = element_text(angle = 90),
              strip.text.x = element_blank())

# combine the plots (and make the legend order consistent)
(p1 / p2) + plot_anno + plot_layout(heights = c(2, 1)) &
  # we have to redo the titles since they change when the legends are altered
  guides(size = guide_legend(title = "Percent Expressed", order = 1),
         color = guide_colorbar(title = "Average Expression", order = 2)) &
  labs(title = NULL, subtitle = NULL, x = NULL, y = NULL)

# save for SDV spreadsheet
sdv$Supplemental_Figure_2A <-
  p1$data %>%
  arrange(id) %>%
  relocate(id, Cell_Type_Full, features.plot, pct.exp, .before = avg.exp) %>%
  rename(seurat_subclusters = id)
sdv$Supplemental_Figure_2B <-
  p2$data %>%
  arrange(id) %>%
  relocate(id, Cell_Type_Full, features.plot, pct.exp, .before = avg.exp) %>%
  rename(annotated_subclusters = id)
```

## Supplemental Figure 3 (IFNG/IL10/FOXP3/TRAIL)

TNFSF10 = TRAIL

```{r figure-supplemental-3-skin-t-feature-nebulosa-vln}
#| fig.dim = c(16, 4.9 * 4)

# making p3 automatically be the same height as p1 and p2 requires getting rid
# of clean_umap, which then introduces different problems

plots_figure_7 <- list()
specific_features <- c("IL10", "FOXP3", "TNFSF10", "IFNG")

for (feature in specific_features) {
  p1 <- FeaturePlot(combined_skin_t, features = feature, pt.size = 0.2,
                    order = TRUE) +
          scale_color_viridis(option = "G", direction = -1) +
          labs(color = "Expression") +
          labels_standard + clean_umap
  p2 <- plot_density(combined_skin_t, features = feature, size = 0.2) +
          labels_standard + clean_umap
  p3 <- vln_cell_info(seurat_obj = combined_skin_t, feature = feature,
                      clrs_specific = colors_annotated_skin_t,
                      split_by = NULL, num_dec = 1, label_size = 3.5) +
          # match labels_standard
          labels_manuscript_violin
  
  # make the density plots match the axes thickness of p1 and p3
  p2[["theme"]][["axis.line"]][["linewidth"]] <-
    p1[[1]][["theme"]][["axis.line"]][["linewidth"]]
  
  # makes the tags align (or at least be closer)
  p1 <- p1[[1]]
  p3 <- p3[[1]]
  
  plots_figure_7[[feature]] <-
    (p1 | p2 | p3) + plot_layout(widths = c(1, 1, 3))
}

wrap_plots(plots_figure_7, nrow = length(specific_features)) +
  plot_anno & labels_manuscript
```

## Supplemental Figure 4 (IFNG/GZMK/PRF1)

```{r figure-supplemental-4-ifng-gzmk-prf1-venn}
#| fig.dim = c(15, 10)

plots_venn_cytotoxic <- list()

# defined in figure 3
for (cell_type in specific_cell_types_skin_gzmk) {
  seurat_obj_cell_type <-
    subset(combined_skin_t, subset = annotated_subclusters == cell_type)
  
  for (sample_type in c("EM", "Control")) {
    seurat_obj <-
      subset(seurat_obj_cell_type, subset = SampleType == sample_type)
    
    cells_gzmb_pos <- GetAssayData(seurat_obj, slot = "scale.data")["GZMB", ]
    cells_gzmb_pos <- cells_gzmb_pos[cells_gzmb_pos > 0] %>% names()
    
    cells_prf1_pos <- GetAssayData(seurat_obj, slot = "scale.data")["PRF1", ]
    cells_prf1_pos <- cells_prf1_pos[cells_prf1_pos > 0] %>% names()
    
    cells_ifng_pos <- GetAssayData(seurat_obj, slot = "scale.data")["IFNG", ]
    cells_ifng_pos <- cells_ifng_pos[cells_ifng_pos > 0] %>% names()
    
    list_venn <- list(GZMB = cells_gzmb_pos, IFNG = cells_ifng_pos,
                      PRF1 = cells_prf1_pos)
    
    plots_venn_cytotoxic[[paste(sample_type, cell_type)]] <-
      ggVennDiagram(x = list_venn,
                    set_color = colors_sample_types_manuscript[[sample_type]],
                    set_size = 0, label_alpha = 0) +
        geom_text(data = venn_setlabel(process_data(Venn(list_venn))),
                aes(X, Y, label = name, size = 5), show.legend = FALSE) +
        labs(title = paste("Skin", cell_type),
             subtitle = paste(sample_type, "only"),
             fill = "Cell Count") +
        scale_fill_gradient2(low = "white", high = "grey50") +
        labels_standard_venn
  }
}

seurat_obj <- combined_skin_t
cell_type <- "CD8+ GZMK+ IFNG T Cells"
seurat_obj$annotated_subclusters <-
  as.character(seurat_obj$annotated_subclusters)
seurat_obj$annotated_subclusters <-
  ifelse(seurat_obj$annotated_subclusters %in% specific_cell_types_skin_gzmk,
         cell_type, seurat_obj$annotated_subclusters)

seurat_obj_cell_type <-
  subset(seurat_obj,
         subset = annotated_subclusters == cell_type)
  
for (sample_type in c("EM", "Control")) {
  seurat_obj <-
    subset(seurat_obj_cell_type, subset = SampleType == sample_type)
  
  cells_gzmb_pos <- GetAssayData(seurat_obj, slot = "scale.data")["GZMB", ]
  cells_gzmb_pos <- cells_gzmb_pos[cells_gzmb_pos > 0] %>% names()
  
  cells_prf1_pos <- GetAssayData(seurat_obj, slot = "scale.data")["PRF1", ]
  cells_prf1_pos <- cells_prf1_pos[cells_prf1_pos > 0] %>% names()
  
  cells_ifng_pos <- GetAssayData(seurat_obj, slot = "scale.data")["IFNG", ]
  cells_ifng_pos <- cells_ifng_pos[cells_ifng_pos > 0] %>% names()
  
  list_venn <- list(GZMB = cells_gzmb_pos, IFNG = cells_ifng_pos,
                    PRF1 = cells_prf1_pos)
  
  plots_venn_cytotoxic[[paste(sample_type, cell_type)]] <-
    ggVennDiagram(x = list_venn,
                  set_color = colors_sample_types_manuscript[[sample_type]],
                  set_size = 0, label_alpha = 0) +
      geom_text(data = venn_setlabel(process_data(Venn(list_venn))),
                aes(X, Y, label = name, size = 5), show.legend = FALSE) +
      labs(title = paste("Skin", cell_type),
           subtitle = paste(sample_type, "only"),
           fill = "Cell Count") +
      scale_fill_gradient2(low = "white", high = "grey50") +
      labels_standard_venn
}

wrap_plots(plots_venn_cytotoxic, nrow = 2, byrow = FALSE) + 
  plot_anno & labels_manuscript_venn
```

## Supplemental Figure 5 (Splicing)

```{r figure-supplemental-5-splicing-umap-dot}
#| fig.dim = c(15, 15)

cd45_feats <- str_c("PTPRC-R", c("A", "B", "C", "O"))

p1 <- UMAPPlot(combined_skin_t_ideis, cols = colors_annotated_skin_t,
               pt.size = 0.2) +
        labs(title = "Skin T Cells") +
        labels_standard + clean_umap +
        theme(legend.text = element_text(size = 11))

# flip the cell type order so it's not upside down
Idents(combined_skin_t_ideis) <-
  factor(combined_skin_t_ideis$annotated_subclusters,
         levels = rev(levels(combined_skin_t_ideis$annotated_subclusters)))

p2 <- DotPlot(combined_skin_t_ideis, features = cd45_feats, dot.scale = 3) %>%
        dot_color_scale() +
        labs(title = "CD45 Isoforms") +
        labels_standard + labels_manuscript_dot

# reflip the order
Idents(combined_skin_t_ideis) <-
  factor(combined_skin_t_ideis$annotated_subclusters,
         levels = levels(combined_skin_t_ideis$annotated_subclusters))

plots_sup_figure_5 <- list()
for (feature in cd45_feats) {
  p3 <- FeaturePlot(combined_skin_t_ideis, features = feature, pt.size = 0.2,
                    order = TRUE) +
          scale_color_viridis(option = "G", direction = -1) +
          labs(color = "Expression") +
          labels_standard + clean_umap#2
  p4 <- plot_density(combined_skin_t_ideis, features = feature, size = 0.2) +
          labels_standard + clean_umap
  
  # make the density plots match the axes thickness of p3
  p4[["theme"]][["axis.line"]][["linewidth"]] <-
    p3[[1]][["theme"]][["axis.line"]][["linewidth"]]
  
  # makes the tags align (or at least be closer)
  p3 <- p3[[1]]
  
  plots_sup_figure_5[[paste0("first_", feature)]] <- p3 # feature
  plots_sup_figure_5[[paste0("second_", feature)]] <- p4 # density
}
# change the order to be more clear
plots_sup_figure_5 <-
  plots_sup_figure_5[sort(names(plots_sup_figure_5))]

# do a box plot instead of a violin plot actually
p5 <- vln_cell_info(seurat_obj = combined_skin_t_ideis, feature = "PTPRC-RA",
                    clrs_specific = colors_annotated_skin_t, split_by = NULL,
                    pt_size = 0.2, num_dec = 1, label_size = 3.5) +
        geom_boxplot(outlier.shape = NA) +
        labels_manuscript_violin
# hide the violin
p5[[1]][["layers"]][[1]] <- NULL

# combine the plots
# 5J's legend is a little off for some reason, probably the fixed aspect ratio
(((p1 + p2) + plot_layout(nrow = 1)) /
    wrap_plots(plots_sup_figure_5, nrow = 2) /
    free(p5, side = "r")) +
  plot_layout(heights = c(1, 2, 1)) + plot_anno & labels_manuscript
```

## Supplemental Figure 6 (Clonal expansion)

By subject:

```{r figure-supplemental-6-clonal-expansion}
#| fig.dim = c(15, 8)

# for sanity checks:
# clone_sizes <- countClones(combined_skin_t[[]],
#                            groups = c("annotated_subclusters", "SampleType"),
#                            clone = "clone_id_unique", remove_na = TRUE)

# only keep the subjects with both EM and control
seurat_obj <- combined_skin_t
overlapping_subjects <- seurat_obj[[]] %>%
                          group_by(Subject) %>%
                          filter(n_distinct(SampleType) == 2) %>%
                          pull(Subject) %>%
                          unique()
seurat_obj <- subset(seurat_obj, Subject %in% overlapping_subjects)

# run APOTC
combined_skin_t_control <-
  RunAPOTC(seurat_obj = subset(seurat_obj, subset = SampleType == "Control"),
           alt_ident = "Subject",
           run_id = "run", clonecall = "clone_id_unique",
           clone_scale_factor = 0.05, rad_scale_factor = 0.8,
           verbose = FALSE)
combined_skin_t_em <-
  RunAPOTC(seurat_obj = subset(seurat_obj, subset = SampleType == "EM"),
           alt_ident = "Subject",
           run_id = "run", clonecall = "clone_id_unique",
           clone_scale_factor = 0.05, rad_scale_factor = 0.8,
           verbose = FALSE)

# the control has fewer subjects than the EM (EM has everything)
# assumes all cell types have at least one clone in both sample types
n_clust <- n_distinct(seurat_obj$Subject)
# adjust the colors and centroids (so that control aligns with EM)
combined_skin_t_control <-
  AdjustAPOTC(combined_skin_t_control, run_id = "run",
              recolor_cluster = 1:n_clust,
              new_color = colors_subject[overlapping_subjects],
              relocate_cluster = 1:n_clust,
              relocation_coord =
                combined_skin_t_em@misc$APackOfTheClones$run@centroids,
              verbose = FALSE)
combined_skin_t_em <-
  AdjustAPOTC(combined_skin_t_em, run_id = "run",
              recolor_cluster = 1:n_clust,
              new_color = colors_subject[overlapping_subjects],
              verbose = FALSE)

# plot the APTOCs
p1 <- APOTCPlot(combined_skin_t_control, run_id = "run",
                retain_axis_scales = TRUE, verbose = FALSE) +
        labs(title = "Control") + labels_standard
p2 <- APOTCPlot(combined_skin_t_em, run_id = "run",
                retain_axis_scales = TRUE, verbose = FALSE) +
        labs(title = "EM") + labels_standard

# reset the axes to be the same
x_limits_p1 <- layer_scales(p1)$x$range$range
x_limits_p2 <- layer_scales(p2)$x$range$range
y_limits_p1 <- layer_scales(p1)$y$range$range
y_limits_p2 <- layer_scales(p2)$y$range$range

x_limits_new <-
  c(min(x_limits_p1[1], x_limits_p2[1]), max(x_limits_p1[2], x_limits_p2[2]))
y_limits_new <-
  c(min(y_limits_p1[1], y_limits_p2[1]), max(y_limits_p1[2], y_limits_p2[2]))

p1 <- p1 + scale_x_continuous(limits = x_limits_new) +
           scale_y_continuous(limits = y_limits_new)
p2 <- p2 + scale_x_continuous(limits = x_limits_new) +
           scale_y_continuous(limits = y_limits_new)

# standardize the legend position
# it will still be slightly different since it plots by the smallest clone and
# not the largest
p1 <- overlayLegend(apotc_ggplot = p1, 
                    legend_position =
                      # c(x_limits_new[2] - 2.5, y_limits_new[2] - 1),
                      c(x_limits_new[1] + 0.5, y_limits_new[1] + 2.5),
                    legend_buffer = 0, legend_label = "Clone Sizes",
                    legend_text_size = 4, add_legend_background = FALSE,
                    add_legend_centerspace = 0)
p2 <- overlayLegend(apotc_ggplot = p2, 
                    legend_position =
                      # c(x_limits_new[2] - 2.5, y_limits_new[2] - 1),
                      c(x_limits_new[1] + 0.5, y_limits_new[1] + 2.5),
                    legend_buffer = 0, legend_label = "Clone Sizes",
                    legend_text_size = 4, add_legend_background = FALSE,
                    add_legend_centerspace = 0)

# combine the plots
(p1 | p2) +
  plot_anno & labels_manuscript_text_big &
  # can't do clean_umap because it will distort the circles
  theme(axis.text.x = element_blank(), axis.text.y = element_blank(),
        axis.ticks = element_blank())
```

Just to get the subject legend for the APOTC plots:

```{r figure-supplemental-6-legend}
#| fig.dim = c(8, 8),
#| eval = FALSE

# only paired subjects
plot_umap(seurat_obj =
            subset(combined_skin_t, subset = SampleType == "Control"),
          tissue_type = "Skin T Cells",
          clrs_specific = colors_subject,
          plot_by = "cluster_all", plot_label = FALSE,
          clusters_col = "Subject") +
  labs(color = "Participant") & labels_manuscript
```

## Supplemental Figure 7 (Clonal Venns)

```{r}
plots_venn_blood_t <- list()

for (cell_type in sort(unique(combined_blood_t$annotated_subclusters))) {
  combined_blood_t_sub <-
    subset(combined_blood_t, subset = annotated_subclusters == cell_type)
  
  # pull clones and remove NAs
  clones_blood <-
    combined_blood_t_sub[[]] %>%
      filter(!is.na(clone_id_unique)) %>%
      pull(clone_id_unique)
  # we only want the overlapping clones (because it makes no sense to show skin
  # clones associated with a blood cell type)
  clones_control <-
    combined_skin_t[[]] %>%
      filter(clone_id_unique %in% clones_blood, SampleType == "Control") %>%
      pull(clone_id_unique)
  clones_em <-
    combined_skin_t[[]] %>%
      filter(clone_id_unique %in% clones_blood, SampleType == "EM") %>%
      pull(clone_id_unique)
  
  list_venn <-
    list("Blood" = clones_blood, "Control" = clones_control, "EM" = clones_em)

  plots_venn_blood_t[[cell_type]] <-
    ggVennDiagram(x = list_venn,
                  set_color = colors_sample_types_manuscript, set_size = 0,
                  label = "count", label_alpha = 0, label_size = 5,
                  edge_size = 1.5) +
      geom_text(data = venn_setlabel(process_data(Venn(list_venn))),
                aes(X, Y, label = name, size = 5), show.legend = FALSE) +
      labs(title = paste("Blood", cell_type, "Clones"),
           fill = "Clone Count") +
      scale_fill_gradient2(low = "white", high = "grey50") +
      # don't cut off "Control"
      scale_x_continuous(expand = expansion(mult = 0.2)) +
      labels_standard_venn +
      theme(plot.title = element_text(size = 14))
}
```

```{r figure-supplemental-7-clonal-venn}
#| fig.dim = c(19, 14)

# try to space them out evenly
wrap_plots(c(plots_venn_blood_t, plots_venn_skin_t), ncol = 4) +
  plot_anno & labels_manuscript_venn
```

## Supplemental Figure 8 (Full chemokines)

Note that:

- CCL6 has been discontinued by HGNC (it's CCL23 now)
- CCL9 is CCL15 in humans
- CCL12 doesn't exist in humans
- CXCL4 is PF4
- CXCL7 is NAP-2/PPBP
- CXCL8	is IL-8 (sometimes)
- There are four atypical chemokine receptors (AKCR[1:4])
- CCRL2 is also atypical

```{r}
# ligands and receptors together
genes_list_ligands <- grep("^(CCL|CX3CL|CXCL|XCL)",
                           rownames(combined_skin_recombined), value = TRUE)
genes_list_receptors <- grep("^(CCR|CX3CR|CXCR|XCR)",
                             rownames(combined_skin_recombined), value = TRUE)

# add alternate gene names
genes_list_ligands <- c(genes_list_ligands, "PF4", "PPBP")

# remove "like" and antisense genes
genes_list_ligands <- setdiff(genes_list_ligands, c("CCL3L1", "CCL4L2"))
genes_list_receptors <- setdiff(genes_list_receptors, "CCR5AS")
```

Combining them makes the figure too tall, so we'll split it up by type:

```{r figure-supplemental-8-chemokines-ligands-heatmap}
#| fig.dim = c(9, 15.75)

chemokine_groups <-
  data.frame("Type" = rep("Ligands", length(genes_list_ligands)),
             "Genes" = genes_list_ligands)
heatmap_DESeq2(DESeq2_out = de_skin_recombined,
               gene_list = genes_list_ligands,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types,
               clrs_anno_top = colors_immune,
               split_row = chemokine_groups,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

```{r figure-supplemental-8-chemokines-receptors-heatmap}
#| fig.dim = c(9, 9.25)

chemokine_groups <-
  data.frame("Type" = rep("Receptors", length(genes_list_receptors)),
             "Genes" = genes_list_receptors)
heatmap_DESeq2(DESeq2_out = de_skin_recombined,
               gene_list = genes_list_receptors,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types,
               clrs_anno_top = colors_immune,
               split_row = chemokine_groups,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

Save for SDV spreadsheet:

```{r}
# pulls from the first lines of heatmap_DESeq2() 
sdv$Supplemental_Figure_8 <-
  enframe(de_skin_recombined) %>%
  unnest(value) %>%
  rename(annotated_clusters_expanded = name) %>%
  dplyr::filter(gene %in% c(genes_list_ligands, genes_list_receptors)) %>%
  # need the genes to be alphanumerically sorted
  mutate(gene = factor(gene,
                       levels = str_sort(unique(gene), numeric = TRUE))) %>%
  arrange(cell_type, gene)
```

## Supplemental Figure 9 (CellChat Heatmap)

```{r figure-supplemental-9-cellchat-heatmap}
#| fig.dim = c(12, 12)

# just go alphabetically and change the styling
p <- netVisual_heatmap(object = cellchat_merged, measure = "weight",
                       color.use = colors_cellchat,
                       # since red and blue are used for the DESeq2 heatmaps
                       color.heatmap = c("#43bb7d", "#b2185b"),
                       title = "Differential Interaction Strength",
                       font.size = 12, font.size.title = 14)

# make NAs gray and add internal lines
p@matrix_color_mapping@na_col <- "grey70"
p@matrix_param[["gp"]][["col"]] <- "black"

# increase the legend size
p@matrix_legend_param$title_gp$fontsize <- 12
p@matrix_legend_param$labels_gp$fontsize <- 10

# increase the bar plot axis sizes
p@top_annotation@anno_list$Strength@fun@var_env$axis_param$gp$fontsize <- 12
p@right_annotation@anno_list$Strength@fun@var_env$axis_param$gp$fontsize <- 15

# change up the labels
p@row_title <- "Senders"

# group the columns like in the other heatmaps
heatmap_order <- skin_recombined_types %>%
                   rownames_to_column(var = "CellType") %>%
                   arrange(CellType) %>%
                   mutate(order = row_number()) %>%
                   arrange(`Immune Cell Type`) %>%
                   pull(order)
p@column_order <- heatmap_order
p@row_order <- heatmap_order

# increase the margin so the cell names aren't cut off and plot
draw(p, padding = unit(c(12.5, 9, 2, 1), "mm")) # bottom, left, top and right
```

```{r}
sdv$Supplemental_Figure_9 <-
  reshape2::melt(data = p@matrix, varnames = c("Senders", "Receivers"),
                 as.is = TRUE,
                 value.name = "Differential Interaction Strength") %>%
  arrange(Senders, Receivers)
```

## Supplemental Figure 10 (ISGs)

```{r figure-supplemental-10-isgs-heatmap}
#| fig.dim = c(9, 14)

# STAT1, ISG15, IRF1, IFITM1
genes_list_marques <- c("GBP1", "IFI35", "IFI6", "IFIT1", "IFIT3", 
                        "MX1", "OAS1", "PSMB8", "TAP1",
                        "IDO1", "KMO", "KYNU")

# significant only, without ones on the previous lists
genes_list_schoggins <- c("ADAR1", "BST2", "CNP", "EIF2AK2", "GBP5", "IFI16",
                          "IFITM3", "LY6E", "MCOLN2", "NCOA7", "RSAD2",
                          "SAT1", "SOCS1", "ZC3HAV1")

# significant only, overlaps a lot with Schoggins
# PML = TRIM19, TMEM173 = STING
genes_list_schneider <- c("PML", "TMEM173")
# "USP18", "TRIM25", "TRIM56" were not significant

# TNFSF10 = TRAIL
genes_list_other <- c("IFITM5", "IFITM10", "STAT4", "STAT5A", "STAT5B", "STAT6",
                      "TNFRSF14", "TNFSF10", "JAK1")

genes_list <- c(genes_list_marques, genes_list_schoggins, genes_list_schneider,
                genes_list_other)

heatmap_DESeq2(DESeq2_out = de_skin_recombined, gene_list = genes_list,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types,
               clrs_anno_top = colors_immune,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

Save for SDV spreadsheet:

```{r}
# pulls from the first lines of heatmap_DESeq2() 
sdv$Supplemental_Figure_10 <-
  enframe(de_skin_recombined) %>%
  unnest(value) %>%
  rename(annotated_clusters_expanded = name) %>%
  dplyr::filter(gene %in% genes_list) %>%
  # need the genes to be alphanumerically sorted
  mutate(gene = factor(gene,
                       levels = str_sort(unique(gene), numeric = TRUE))) %>%
  arrange(annotated_clusters_expanded, gene)
```

## Supplemental Figure 11 (Full interleukins)

The full list has 101 hits going off "IL*". [HUGO says](https://www.genenames.org/data/genegroup/#!/group/601) there are 43 IL genes.

The two receptor agonists (IL1RN and IL46RN) have been excluded as they don't have any intracellular effect.

```{r figure-supplemental-11-il-heatmap}
#| fig.dim = c(9, 16)

# CSV download from https://www.genenames.org/data/genegroup/#!/group/601
# il_genes <- read_csv("~/Downloads/group-601.csv", skip = 1,
#                      show_col_types = FALSE)
# then I added a few manually

genes_list <- c(paste0("IL", 1:36),
                "IL1F10", 
                "IL1A", "IL1B", "IL12A", "IL12B", "IL17A", "IL17B", "IL17C",
                "IL17D", "IL17F", "IL23A",
                "CXCL8", # CXCL8 is IL8
                "TXLNA", # TXLNA is IL14
                "IFNL2", # IFNL2 is IL28A
                "IFNL3", # IFNL3 is IL28B
                "IFNL1", # IFNL1 is IL29
                "IL36A", "IL36B", "IL36G", "IL37")
# Have an A/B: IL1, IL12, IL17, IL23, IL36
# Don't exist: IL8, IL14, IL28, IL29, IL30 (is now IL27), IL35 (EBI3 + IL12A)

heatmap_DESeq2(DESeq2_out = de_skin_recombined, gene_list = genes_list,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types,
               clrs_anno_top = colors_immune,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

Save for SDV spreadsheet:

```{r}
# pulls from the first lines of heatmap_DESeq2() 
sdv$Supplemental_Figure_11 <-
  enframe(de_skin_recombined) %>%
  unnest(value) %>%
  rename(annotated_clusters_expanded = name) %>%
  dplyr::filter(gene %in% genes_list) %>%
  # need the genes to be alphanumerically sorted
  mutate(gene = factor(gene,
                       levels = str_sort(unique(gene), numeric = TRUE))) %>%
  arrange(annotated_clusters_expanded, gene)
```

## Supplemental Figure 12 (Metabolism)

Get the genes from our two pathways of interest:

```{r}
genes_list_gly <-
  filter(hallmark, gs_name == "HALLMARK_GLYCOLYSIS")$gene_symbol
genes_list_gly <-
  intersect(genes_list_gly, unique(de_genes_skin_recombined$gene))
sort(genes_list_gly)

genes_list_op <-
  filter(hallmark, gs_name == "HALLMARK_OXIDATIVE_PHOSPHORYLATION")$gene_symbol
genes_list_op <-
  intersect(genes_list_op, unique(de_genes_skin_recombined$gene))
sort(genes_list_op)

# there is some overlap between pathways
length(intersect(genes_list_gly, genes_list_op))
```

Group the genes:

```{r}
# NADH:Ubiquinone oxidoreductase subunits
genes_complex_I <- c("NDUFA1", "NDUFA2", "NDUFA3", "NDUFA4", "NDUFA5", "NDUFA6",
                     "NDUFA7", "NDUFA8", "NDUFA9", "NDUFAB1", "NDUFB2",
                     "NDUFB3", "NDUFB4", "NDUFB5", "NDUFB6", "NDUFB7", "NDUFB8",
                     "NDUFC1", "NDUFC2", "NDUFS2", "NDUFS3", "NDUFS4", "NDUFS6",
                     "NDUFS7", "NDUFS8", "NDUFV1", "NDUFV2")
# Succinate dehydrogenase subunits
genes_complex_II <- c("SDHA", "SDHC", "SDHD")
# Ubiquinol-Cytochrome c Reductase subunits
genes_complex_III <- c("UQCR10", "UQCR11", "UQCRB", "UQCRC2", "UQCRFS1",
                       "UQCRH", "UQCRQ")
# Cytochrome c oxidase subunits
genes_complex_IV_1 <- c("COX4I1", "COX5A", "COX5B", "COX6A1")
genes_complex_IV_2 <- c("COX6B1", "COX6C", "COX7A2", "COX7A2L", "COX7B",
                        "COX7C", "COX8A", "COX17")
# ATP Synthase subunits
genes_complex_V <- c("ATP5F1A", "ATP5F1B", "ATP5F1C", "ATP5F1D", "ATP5F1E",
                     "ATP5MC1", "ATP5MC2", "ATP5MC3", "ATP5ME", "ATP5MF",
                     "ATP5MG", "ATP5PB", "ATP5PD", "ATP5PF", "ATP5PO")

# ERP29 is really for ER (endoplasmic reticulum stress) and not oxidative stress
genes_apoptosis_stress <- c("ERP29", "LGALS1", "BAX", "IDH1", "IDH2")
genes_glycolysis <- c("ENO1", "HK2", "LDHA", "LDHB", "MIF", "PDHA1", "PYGB")
genes_mito_metabolism <- c("DLD", "ECH1", "ECHS1", "ETFDH", "HADHB", "HCCS",
                           "HSD17B10", "MDH1", "MDH2", "SDC2", "SLC25A3",
                           "SLC25A5", "SLC25A6", "SUCLG1", "VDAC1")

# combine the groups
genes_list_part1 <- c(genes_apoptosis_stress, genes_complex_I, genes_complex_II,
                      genes_complex_III, genes_complex_IV_1)
genes_list_part2 <- c(genes_complex_IV_2, genes_complex_V, genes_glycolysis,
                      genes_mito_metabolism)

# check that the lengths are as close as possible
length(genes_list_part1)
length(genes_list_part2)
```

```{r figure-supplemental-12-metabolism-part1-heatmap}
#| fig.dim = c(9, 17)

part1_groups <-
  data.frame("Type" = c(rep("Apoptosis & Stress", length(genes_apoptosis_stress)),
                        rep("Complex I", length(genes_complex_I)),
                        rep("Complex II", length(genes_complex_II)),
                        rep("Complex III", length(genes_complex_III)),
                        rep("Complex IV", length(genes_complex_IV_1))),
             "Genes" = genes_list_part1)

heatmap_DESeq2(DESeq2_out = de_skin_recombined,
               gene_list = genes_list_part1,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types,
               clrs_anno_top = colors_immune,
               split_row = part1_groups,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

```{r figure-supplemental-12-metabolism-part2-heatmap}
#| fig.dim = c(9, 17)

part2_groups <-
  data.frame("Type" = c(rep("Complex IV", length(genes_complex_IV_2)),
                        rep("Complex V", length(genes_complex_V)),
                        rep("Glycolysis", length(genes_glycolysis)),
                        rep("Mitochondrial Metabolism", length(genes_mito_metabolism))),
             "Genes" = genes_list_part2)

heatmap_DESeq2(DESeq2_out = de_skin_recombined,
               gene_list = genes_list_part2,
               FDR_cutoff = fdr_cutoff,
               anno_top = skin_recombined_types,
               clrs_anno_top = colors_immune,
               split_row = part2_groups,
               x_angle = 90, row_size = 14, col_size = 14, number_size = 16)
```

Save for SDV spreadsheet:

```{r}
# pulls from the first lines of heatmap_DESeq2() 
sdv$Supplemental_Figure_12 <-
  enframe(de_skin_recombined) %>%
  unnest(value) %>%
  rename(annotated_clusters_expanded = name) %>%
  dplyr::filter(gene %in% c(genes_list_part1, genes_list_part2)) %>%
  # need the genes to be alphanumerically sorted
  mutate(gene = factor(gene,
                       levels = str_sort(unique(gene), numeric = TRUE))) %>%
  arrange(annotated_clusters_expanded, gene)
```

## Supplemental Figure 13 (Aggregation)

```{r figure-supplemental-13-aggr-read-counts-bar}
#| fig.dim = c(12, 8)

# Read Counts for All Samples and GEX Unaggregated Data
p1 <- (ggplot(metrics_summary,
              aes(x = SampleName, y = NumberofReads, fill = Dataset)) +
          geom_col(color = "black", linewidth = 0.2) +
          labs(x = "Sample", y = "Number of Reads (Pre Aggregation)") +
          scale_fill_manual(values = unname(colors_dataset)) +
          theme_bw + labels_standard + labels_rotate_x +
          labels_manuscript_text +
          theme(legend.position = "none")) %>% 
        add_info_bar(info_type = "Dataset", text_size = 11)

# Read Counts per Cell for All Samples and GEX Unaggregated Data
p2 <- (ggplot(metrics_summary %>%
                mutate(min_rpc = SampleName == "192565SKL",
                       label = ifelse(min_rpc, "", "")),
              aes(x = SampleName, y = ReadsPerCell, fill = Dataset)) +
         geom_col(color = "black", linewidth = 0.2) +
         geom_text(aes(label = label), vjust = -0.3, size = 6) +
         labs(x = "Sample", y = "Reads per Cell (Pre Aggregation)") +
         scale_fill_manual(values = unname(colors_dataset)) +
         theme_bw + labels_standard + labels_rotate_x +
         labels_manuscript_text + theme(legend.position = "none")) %>%
         add_info_bar(info_type = "Dataset", text_size = 11)

(p1 / p2) + plot_anno + plot_layout(axes = "collect")

# save for SDV spreadsheet
sdv$Supplemental_Figure_13 <- metrics_summary %>% select(-Timepoint)
```

## Supplemental Figure 14 (Blood UMAPs)

```{r figure-supplemental-14-blood-umaps}
#| fig.dim = c(16, 14)

p1 <- plot_umap(seurat_obj = combined_blood, tissue_type = "Blood",
                use_hues = TRUE, plot_by = "cluster_all", label_size = 3.5) +
        guides(color = guide_legend(override.aes = list(size = 3), ncol = 2))

p2 <- plot_umap(seurat_obj = combined_blood, tissue_type = "Blood",
                clrs_specific = colors_annotated_blood,
                plot_by = "cluster_all", label_size = 3.5, annotated = TRUE)

p3 <- plot_umap(seurat_obj = combined_blood_t, tissue_type = "Blood T Cells",
                use_hues = TRUE, plot_by = "cluster_all", label_size = 3.5,
                clusters_col = "seurat_subclusters")

p4 <- plot_umap(seurat_obj = combined_blood_t, tissue_type = "Blood T Cells",
                clrs_specific = colors_annotated_blood_t,
                plot_by = "cluster_all", label_size = 3.5, annotated = TRUE,
                annotations_col = "annotated_subclusters")

((p1 + p2 + p3 + p4) & labels_manuscript_text_big) +
  plot_anno + plot_layout(nrow = 2) 
```

------------------------------------------------------------------------

# Reviewer Figures

We included these in our response to reviewers, but not the manuscript itself.

## Batch effects

```{r figure-reviewer-batch-effects-umap}
#| fig.dim = c(16, 8)

# by subject
p1 <- UMAPPlot(object = combined_skin, cols = colors_subject,
               pt.size = 0.1, group.by = "Subject", label = FALSE,
               shuffle = TRUE, seed = 42, raster = FALSE) +
        labs(title = NULL, color = "Subject") +
        labels_standard + clean_umap + labels_manuscript_text

# by dataset
p2 <- UMAPPlot(object = combined_skin, cols = colors_dataset_manuscript,
               pt.size = 0.1, group.by = "Dataset", label = FALSE,
               shuffle = TRUE, seed = 42, raster = FALSE) +
        labs(title = NULL, color = "Dataset") +
        labels_standard + clean_umap + labels_manuscript_text

p1 | p2
```

## TCR overlaps

> "Supplemental data showing TCRa/b chain pairing would be helpful to include."

Read in the raw data (because `combined_tcr` already has the TRA-only chains filtered out):

```{r}
# using the code from 05-quality_control.Rmd
barcodes_TCR_beta <- list()
barcodes_TCR_alpha <- list()

for (sample in samples) {
  # setup
  meta_sample <- dplyr::filter(meta, SampleName == sample)
  path_data <- file.path(path_data_root,
                         paste0("dataset", unique(meta_sample$Dataset)),
                         "default")
  
  # read in the data
  data_TCR <-
    read_csv(file.path(path_data,
                       filter(meta_sample, DataType == "TCR")$SampleDir, "outs",
                       "filtered_contig_annotations.csv"),
             show_col_types = FALSE)

  barcodes_TCR_beta[[sample]] <-
      reformat_vdj_barcode_sample(filter(data_TCR, chain == "TRB"),
                                  sample_name = sample)
  barcodes_TCR_alpha[[sample]] <-
      reformat_vdj_barcode_sample(filter(data_TCR, chain == "TRA"),
                                  sample_name = sample)
}
```

Summarize the information:

```{r}
TCR_summary <- c()

for (sample in samples) {
  barcode_TRA <- barcodes_TCR_alpha[[sample]]
  barcode_TRB <- barcodes_TCR_beta[[sample]]
  barcodes_shared <- unique(intersect(barcode_TRA, barcode_TRB))

  barcode_TRA_only <-
    unique(barcode_TRA[which(!barcode_TRA %in% barcodes_shared)])
  barcode_TRB_only <-
    unique(barcode_TRB[which(!barcode_TRB %in% barcodes_shared)])

  TCR_summary <- bind_rows(TCR_summary, c("TRAOnly" = length(barcode_TRA_only),
                                          "TRBOnly" = length(barcode_TRB_only),
                                          "Paired" = length(barcodes_shared)))
}

TCR_summary <- meta %>%
                select(SampleName, Subject, SampleType, Dataset) %>%
                distinct() %>%
                bind_cols(TCR_summary) %>%
                # make sure Dataset is a character
                mutate(Total = rowSums(across(where(is.numeric))))

# for plotting
chain_types_tcr <- c("TRBOnly", "Paired", "TRAOnly")
TCR_summary_df <-
  TCR_summary %>%
  pivot_longer(cols = all_of(chain_types_tcr),
               names_to = "ChainType", values_to = "Count") %>%
  mutate(ChainType = factor(ChainType, levels = chain_types_tcr)) %>%
  # change the Dataset to match the manuscript
  mutate(Dataset = as.integer(Dataset), Dataset = Dataset - 1)
```

Group by dataset:

```{r figure-reviewer-chains-tcr-dataset-bar}
#| fig.dim = c(10, 5)

plot_counts(summary_df = TCR_summary_df %>%
                           group_by(Dataset, ChainType) %>%
                           summarize(Count = sum(Count)),
            data_types = "TCR", count_type = "Barcode",
            fill_type_label = "Chain Type",
            x_axis = "Dataset", x_axis_label = "Dataset",
            clrs_datatype = colors_chain_tcr) +
  labs(title = NULL, y = "Number of Cell Barcodes") +
  coord_flip() +
  labels_manuscript_text +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5),
        panel.grid.minor.x = element_blank())
```

------------------------------------------------------------------------

# Supporting Data Values

```{r}
path_sdv <-
  file.path(path_tables, "manuscript", "supporting_data_values.xlsx")

# check if the spreadsheet already exists
# delete if it does so an error isn't thrown when knitting this document
if (file.exists(path_sdv)) file.remove(path_sdv)

for (df in names(sdv)) {
  write.xlsx(x = as.data.frame(sdv[[df]]), file = path_sdv,
             sheetName = df, col.names = TRUE, row.names = FALSE,
             append = TRUE)
}
```

------------------------------------------------------------------------

# Text

## Alternate gene names

In [common name] / [marker name] order:

* CD16 / FCGR3A
* CD20 / MS4A1
* CD25 / IL2RA
* CD56 / NCAM1
* CD127 / IL7R
* CD132 / IL2RG
* CD225 / IFITM1
* CD301 / CLEC10A
* CD303 / CLEC4C
* CXCL4 / PF4
* HSP27 / HSPB1 & HSPB2
* HSP60 / HSPD1
* HSP70 / HSPA4
* HSP72 / HSPA1A
* IL8 / CXCL8
* IL14 / TXLNA
* IL28A / IFNL2
* IL28B / IFNL3
* IL29 / IFNL1
* PDL1 / CD274
* STING / TMEM173
* TRIF / TICAM1

## Results

### Cell type counts

Blood:

```{r}
# blood
cat(paste("There are", nlevels(combined_blood$annotated_clusters),
          "blood cell types:\n",
          str_c(levels(combined_blood$annotated_clusters),
                collapse = ", ")))

# blood T cells
cat(paste("There are", nlevels(combined_blood_t$annotated_subclusters),
          "blood T cells cell types:\n",
          str_c(levels(combined_blood_t$annotated_subclusters),
                collapse = ", ")))
```

Skin:

```{r}
# skin
cat(paste("There are", nlevels(combined_skin$annotated_clusters),
          "skin cell types:\n",
          str_c(levels(combined_skin$annotated_clusters),
                collapse = ", ")))

# skin (recombined)
cat(paste("There are",
          nlevels(combined_skin_recombined$annotated_clusters_expanded),
          "skin (recombined) cell types:\n",
          str_c(levels(combined_skin_recombined$annotated_clusters_expanded),
                collapse = ", ")))

# skin T cells
cat(paste("There are", nlevels(combined_skin_t$annotated_subclusters),
          "skin T cells cell types:\n",
          str_c(levels(combined_skin_t$annotated_subclusters),
                collapse = ", ")))
```

### Means and SDs

> "An average of 5,882 cells were identified per sample across all of the tissues and there were an estimated 41,568 cells in the blood and 146,662 cells in the (combined EM and control) skin." - Methods

> "In the methods, please provide the SD for means, and the mean/median (IQR) number of cells per group (EM vs control), including after filtering." - Reviewer B

```{r}
all_cells <- metrics_summary$EstimatedNumberofCells

cat(paste("An average of", round(mean(all_cells)),
          "cells were identified per sample across all of the tissues, with a",
          "standard deviation of", round(sd(all_cells)), "cells."))
```

Before filtering:

```{r}
cat(paste("There are", ncol(combined_skin_unprocessed),
          "cells in the initial unprocessed skin data."))

combined_skin_unprocessed[[]] %>%
  group_by(Subject, SampleType) %>%
  summarize(cell_count = n()) %>%
  group_by(SampleType) %>%
  summarize(Mean = round(mean(cell_count)), Median = round(median(cell_count)),
            SD = round(sd(cell_count)))
```

After filtering:

```{r}
cat(paste("There are", ncol(combined_skin), "cells in the filtered skin data."))

combined_skin[[]] %>%
  group_by(Subject, SampleType) %>%
  summarize(cell_count = n()) %>%
  group_by(SampleType) %>%
  summarize(Mean = round(mean(cell_count)), Median = round(median(cell_count)),
            SD = round(sd(cell_count)))
```

### Majority immune types

```{r}
combined_skin[[]] %>%
  group_by(SampleType) %>%
  summarize(Immune_Cells = sum(annotated_clusters_immune == "Immune"),
            Perc_Immune = label_percent(accuracy = 0.1)(Immune_Cells / n()))
```

### Clonal overlaps

#### Counts

Total:

```{r}
print(process_set_data(list_venn_total) %>% select(-id, -item))
print(process_region_data(list_venn_total) %>% select(-id, -item))
```

```{r}
total_blood <-
  unname(filter(process_set_data(list_venn_total), name == "Blood")$count)
total_blood_control <-
  sum(filter(process_region_data(list_venn_total),
             name %in% c("Blood/Control", "Blood/Control/EM"))$count)
total_blood_em <-
  sum(filter(process_region_data(list_venn_total),
             name %in% c("Blood/EM", "Blood/Control/EM"))$count)

cat(paste(label_percent(accuracy = 0.01)(total_blood_control/total_blood),
          "of the total blood T cell clonotypes are shared with the control."))
cat(paste(label_percent(accuracy = 0.01)(total_blood_em/total_blood),
          "of the total blood T cell clonotypes are shared with the EM."))
```

Skin:

```{r}
for (cell_type in specific_cell_types_skin_simpler) {
  print(cell_type)
  
  # set up variables
  specific_venn <- list_venn_skin_t_simpler[[cell_type]]
  venn_set <- process_set_data(specific_venn) %>% select(-id, -item)
  venn_region <- process_region_data(specific_venn) %>% select(-id, -item)
  
  print(venn_set)
  print(venn_region)
  
  count_control <- sum(filter(venn_set, name == "Control")$count)
  count_em <- sum(filter(venn_set, name == "EM")$count)
  count_blood_control <-
    sum(filter(venn_region,
               name %in% c("Blood/Control", "Blood/Control/EM"))$count)
  count_blood_em <-
    sum(filter(venn_region,
               name %in% c("Blood/EM", "Blood/Control/EM"))$count)
  
  print(paste(label_percent(accuracy = 0.01)(count_blood_control/count_control),
              "of the control skin", cell_type,
              "clonotypes are shared with the blood.",
              label_percent(accuracy = 0.01)(count_blood_em/count_em),
              "of the EM skin", cell_type, "clonotypes are shared with the blood."))
  cat("\n")
}

for (cell_type in specific_cell_types_skin) {
  print(cell_type)
  
  # set up variables
  specific_venn <- list_venn_skin_t[[cell_type]]
  venn_set <- process_set_data(specific_venn) %>% select(-id, -item)
  venn_region <- process_region_data(specific_venn) %>% select(-id, -item)
  
  print(venn_set)
  print(venn_region)
  
  count_skin <- sum(filter(venn_set, name == "EM")$count)
  count_blood_em <-
    sum(filter(venn_region,
               name %in% c("Blood/EM", "Blood/Control/EM"))$count)
  
  print(paste(label_percent(accuracy = 0.01)(count_blood_em/count_skin),
            "of the EM", cell_type, "clonotypes are shared with the blood."))
  cat("\n")
}
```

#### Significance

*Note that blood = 1, control = 2, and EM = 3 in the Venn calls.*

Total:

```{r}
venn <- list_venn_total

overlap_blood_em <- length(overlap(venn, c(1, 3)))
overlap_blood_control <- length(overlap(venn, 1:3))
total_em <- length(overlap(venn, 3))
total_control <- length(overlap(venn, 2))
em_no_blood <- length(discern(venn, 3, 1))
control_no_blood <- length(discern(venn, 2, 1))

# two-sample proportions test
result <- prop.test(c(overlap_blood_em, overlap_blood_control),
                    c(total_em, total_control))
print(result)
```

Simpler:

```{r}
for (cell_type in specific_cell_types_skin_simpler) {
  print(cell_type)
  
  venn <- list_venn_skin_t_simpler[[cell_type]]
  
  overlap_blood_em <- length(overlap(venn, c(1, 3)))
  overlap_blood_control <- length(overlap(venn, 1:3))
  total_em <- length(overlap(venn, 3))
  total_control <- length(overlap(venn, 2))
  em_no_blood <- length(discern(venn, 3, 1))
  control_no_blood <- length(discern(venn, 2, 1))
  
  # two-sample proportions test
  result <- prop.test(c(overlap_blood_em, overlap_blood_control),
                      c(total_em, total_control))
  print(result)
}
```

Specific cell types:

```{r}
for (cell_type in specific_cell_types_skin) {
  print(cell_type)
  
  venn <- list_venn_skin_t[[cell_type]]
  
  overlap_blood_em <- length(overlap(venn, c(1, 3)))
  overlap_blood_control <- length(overlap(venn, 1:3))
  total_em <- length(overlap(venn, 3))
  total_control <- length(overlap(venn, 2))
  em_no_blood <- length(discern(venn, 3, 1))
  control_no_blood <- length(discern(venn, 2, 1))
  
  # two-sample proportions test
  result <- prop.test(c(overlap_blood_em, overlap_blood_control),
                      c(total_em, total_control))
  print(result)
}
```

### GSEA

How many of the possible pathways had significance for at least one cell type?

```{r}
cat(paste("Out of the", n_distinct(hallmark$gs_name),
          "total Hallmark pathways,", n_distinct(sig_gsea_simpler$Pathways),
          "had significance for at least one cell type.",
          "The ones that didn't were:",
          toString(setdiff(unique(hallmark$gs_name),
                           unique(sig_gsea_results$ID)) %>%
                     str_remove("HALLMARK_") %>% str_replace_all("_", " "))))
```

How many of the genes in the two selected metabolic pathways are significantly differentially expressed?

```{r}
for (pathway in
     c("HALLMARK_GLYCOLYSIS", "HALLMARK_OXIDATIVE_PHOSPHORYLATION")) {
  genes_list <- filter(hallmark, gs_name == pathway)$gene_symbol
  genes_list <- intersect(genes_list, unique(de_genes_skin_recombined$gene))
  
  cat(paste("There are", length(genes_list), "significant genes in the", pathway,
            "pathway. "))
}
```

## Methods

### Data processing

```{r}
metrics_summary %>%
  group_by(Dataset) %>%
  summarize(NumberofReads = mean(NumberofReads)) %>%
  mutate(NumberofReads = prettyNum(NumberofReads, big.mark = ",")) %>%
  print_kable(kable_height = NULL)

# alternatively
metrics_summary %>%
  group_by(Dataset) %>%
  summarize(NumberofReads = mean(NumberofReads)) %>%
  mutate(NumberofReads = round_any(NumberofReads, accuracy = 10^6)) %>%
  print_kable(kable_height = NULL)
```

### Quality control

Blood:

```{r}
# could also use the number of columns in the unprocessed object if you saved it
cat(paste("There are",
          dplyr::filter(metrics_summary, SampleType == "Blood") %>%
            pull(EstimatedNumberofCells) %>%
            sum() %>%
            prettyNum(big.mark = ","),
          "cells in the initial unprocessed blood data."))

cat(paste("There are",
          ncol(combined_blood_with_doublets) %>% prettyNum(big.mark = ","),
          "total cells in the processed (with doublets) blood."))

cat(paste("There are",
          ncol(combined_blood) %>% prettyNum(big.mark = ","),
          "total cells in the processed (doublets removed) blood data."))
```

Skin:

```{r}
# could also use the number of columns in the unprocessed object if you saved it
cat(paste("There are",
          dplyr::filter(metrics_summary, SampleType != "Blood") %>%
            pull(EstimatedNumberofCells) %>%
            sum() %>%
            prettyNum(big.mark = ","),
          "cells in the initial unprocessed skin data."))

cat(paste("There are",
          ncol(combined_skin_with_doublets) %>% prettyNum(big.mark = ","),
          "total cells in the processed (with doublets) skin data;",
          table(combined_skin_with_doublets$SampleType)[["EM"]] %>%
            prettyNum(big.mark = ","), "from the EMs and",
          table(combined_skin_with_doublets$SampleType)[["Control"]] %>%
            prettyNum(big.mark = ","),
          "from the controls"))

cat(paste("There are",
          ncol(combined_skin) %>% prettyNum(big.mark = ","),
          "total cells in the processed (doublets removed) skin data;",
          table(combined_skin$SampleType)[["EM"]] %>%
            prettyNum(big.mark = ","), "from the EMs and",
          table(combined_skin$SampleType)[["Control"]] %>%
            prettyNum(big.mark = ","),
          "from the controls"))
```

### Averages

Using the metrics summaries (just for the gene expression):

```{r}
cat(paste("An average of", round(mean(metrics_summary$NumberofReads) / 1e6),
          "million reads were sequenced and an average of around",
          round(mean(metrics_summary$EstimatedNumberofCells)),
          "cells were identified per sample."))
```

### Recombination

```{r}
cat(paste(table(combined_skin$annotated_clusters)[["T Cells"]] -
            table(combined_skin_recombined$annotated_clusters)[["T Cells"]],
          "skin T cells were removed after recombination."))
```

### TCR

```{r}
combined_skin_t[[]] %>% group_by(Has_TCR) %>% summarize(n = n())
combined_skin_t[[]] %>% group_by(Has_TCR, paired_alpha) %>% summarize(n = n())
```
