---
title: "Splicing"
output:
  html_document:
    code_download: yes
    code_folding: hide
    css: ../../"style.css"
    dev: "png"
    df_print: kable
    fig_caption: yes
    keep_md: no
    self_contained: true
    theme: yeti
    toc: yes
    toc_depth: 4
    toc_float:
        collapsed: false
        smooth_scroll: false
editor_options:
  chunk_output_type: console
---

------------------------------------------------------------------------

# Overview and setup

```{r setup-options, include = FALSE}
knitr::opts_chunk$set(comment = NA, dpi = 200, fig.align = "center",
                      out.width = "100%", warning = FALSE, error = TRUE,
                      message = FALSE,
                      fig.path = here::here("primary_analysis", "figures",
                                            "downstream", "splicing",
                                            "splicing-"))

# will mess with chunk labels, but needed for parameterized reports
options(knitr.duplicate.label = "allow")
```

**Author(s):** Edel Aron<br>
**Last Updated:** `r Sys.Date()`

**R version:** `r R.Version()$version.string`<br>
**Platform:** `r R.Version()$platform`<br>
**Running under:** `r sessionInfo()$running`

```{r, include = FALSE}
# load packages, colors and themes and functions (R code only)
code_setup <-
  knitr::purl(here::here("single_cell", "primary_analysis", "notebooks",
                         "01-setup.Rmd"),
              output = tempfile(), documentation = 0)
code_functions <-
  knitr::purl(here::here("single_cell", "primary_analysis", "notebooks",
                         "02-functions.Rmd"),
              output = tempfile(), documentation = 0)
code_meta <- 
  knitr::purl(here::here("single_cell", "primary_analysis", "notebooks",
                         "03-meta.Rmd"),
              output = tempfile(), documentation = 0)

# read in the code, then remove the temp files
source(code_setup)
source(code_functions)
source(code_meta)
unlink(c(code_setup, code_functions, code_meta))
```

*I also tried Velocyto, SCASL, and MARVEl but they were difficult to install and/or run.*

------------------------------------------------------------------------

# Read in the data

```{r setup-gex}
# object
combined_skin_t <-
  readRDS(file.path(path_objs, "final", "combined_skin_t_ref.rds"))
combined_skin_t_ideis <-
  readRDS(file.path(path_objs, "downstream", "splicing",
                    "combined_skin_t_ideis.rds"))

# colors
colors_annotated_skin_t <- combined_skin_t@misc$colors_annotated
```

------------------------------------------------------------------------

# Background

Literature:

- [Populations of Human T Lymphocytes That Traverse the Vascular Endothelium Stimulated by Borrelia burgdorferi Are Enriched with Cells That Secrete Gamma Interferon](https://journals.asm.org/doi/full/10.1128/iai.72.3.1530-1536.2004)
  - Consequently, enrichment of type 1 T lymphocytes in the migrated populations is unlikely to be merely secondary to the increased tendency of CD45RO-bearing cells to traverse the endothelium.
- [Differential expression and regulation of cytokine mRNAs in normal human CD45R T cell subsets Author links open overlay panel](https://www.sciencedirect.com/science/article/pii/1043466694900329?via%3Dihub)
  - Both CD45RA+ and CD45 RO+ populations produced mRNAs for interleukin (IL)-2, IL-2 receptor (alpha chain), IL-6 receptor and tumour necrosis factor (TNF)-beta within 3–4 h of activation. Whilst IL-3 and RANTES were also expressed in both subsets, CD45RO+ cells were clearly the major producers of these cytokines. In contrast, mRNA transcripts for IL-1 alpha, IL-4, IL-5, IL-6, IL-10, interferon gamma (IFN-γ) and the T cell receptor for IL-1 were almost exclusively induced in CD45RO+ T cells. A population of CD4+ T cells co-expressing intermediate levels of both CD45RA and CD45RO, namely CD45RA+/CD45RO+, appeared to be the major producers of IL-6.
  - Multiple isoforms of CD45, with molecular weights from about 220 kDa to 180 kDa are generated by the alternate splicing of three exons (4,5 and 6; designated A, B and C), each encoding approximately 50 amino acids within the N-terminal extracellular domain. Thus, eight different transcripts derived from exons, A, B and C are possible (i.e. ABC, AB, AC, BC, A, B, C and one lacking all three, namely RO).
  - Differential isoform expression on T cells appears to correlate with prior antigenic exposure and function. Naive or unprimed CD4+T cells isolated from peripheral blood display the higher molecular weight isoforms while memory cells predominantly express the isoform lacking exons A, B and C. These subsets are referred to as CD45RA+ and CD45RO+ respectively.
  - Accumulating evidence indicates that CD4+CD45RA+ and CD4+CD45RO+ T cells display different functional and migratory properties. For example, CD45RA T cells function as inducers of cytotoxic CD8+ T cells and migrate from blood to lymph nodes via high endothelial venules.
  - CD45RO+ cells, on the other hand, migrate to the lymph via tissues, possess the capacity to respond to recall antigen and provide B cell helper function. Whilst they represent non-overlapping subsets, a proportion of resting peripheral blood T cells express intermediate levels of both CD45RA and CD45RO isoforms on the cell surface.
- [Activation of Endothelium by Borrelia burgdorferi In Vitro Enhances Transmigration of Specific Subsets of T Lymphocytes](https://journals.asm.org/doi/full/10.1128/iai.69.4.2190-2197.2001)
  - Compared to the initially added population of T lymphocytes, the population that migrated across untreated endothelium or HUVEC activated with B. burgdorferi or IL-1 contained a significantly smaller percentage of CD45RA+RO− (naı̈ve) cells and a greater proportion of CD45RA+RO+ cells.

------------------------------------------------------------------------

# IDEIS

## Setup

[GitHub](https://github.com/Lab-of-Adaptive-Immunity/IDEIS) and [paper](https://pmc.ncbi.nlm.nih.gov/articles/PMC11496083/)

* "We showed that IDEIS accurately identifies canonical human CD45 isoforms in datasets generated by 10× Genomics 5’ sequencing assays."
* "Haradhvala et al. previously developed CD45er, a script for the analysis of PTPRC/CD45 splicing in scRNA-seq data to evaluate engineered T cells in B-cell lymphoma therapy (26). A comparison of this tool and IDEIS showed that, while both programs have similar results to human data, IDEIS has a substantial advantage for third-party users. First, IDEIS uses a faster algorithm than CD45er does. Second, IDEIS is much easier to use with murine data than CD45er. Third, the results generated by IDEIS are easier to use, as they are exported as MTX objects that can be loaded in R for downstream analysis. If the user opts to use the Seurat object as an input, the results of the analysis are added directly to it as a new assay, making further analyses even easier."

*Note that this is a Python package.*

```{bash, eval = FALSE}
# To obtain the copy of this software simply clone this repository in your current directory using the following command:
cd ~/git/Papers # change the location as needed
git clone https://github.com/Lab-of-Adaptive-Immunity/IDEIS

# Some modes of software will require 10X Cell Ranger whitelists to be present in 'IDEIS/cr_whitelists' subdirectory, specifically files named '3M-february-2018.txt' and '737K-august-2016.txt'. You can find them in your installation of 10X Cell Ranger at 'cellranger-x.y.z/cellranger-cs/x.y.z/lib/python/cellranger/barcodes' (for versions 3 and earlier) or at 'cellranger-x.y.z/lib/python/cellranger/barcodes/' (for later versions), but you can also download them here:
ls /opt/cellranger-6.1.2/lib/python/cellranger/barcodes # or whatever version you have

# Finally, you can download an example data set to run file on, but it is not required; you may want to skip directly to the analysis of your own data.
mkdir /media/edelaron/Midas/packages/ideis # change the location as needed
cd /media/edelaron/Midas/packages/ideis
git clone https://github.com/Lab-of-Adaptive-Immunity/Test_Data_IDEIS

# We provide a way to create conda environment with these packages installed. The 'Conda' directory provides a file 'conda_IDEIS_env_specifications.txt' that specifies complete environment needed to run IDEIS. If you don't have Conda, install it according to instructions here: https://conda.io/projects/conda/en/latest/user-guide/install/index.html. Then, create and activate Conda environment for IDEIS:
cd ~/git/Papers/IDEIS/
conda create --name IDEIS_env --file Conda/conda_IDEIS_env_specifications.txt
conda activate IDEIS_env

# Once the environment is ready and activated, you can clone IDEIS using the command specified in "Getting IDEIS" section if not done. Lastly, you need to obtain the software bamtofastq-1.4.1. You can get it on this page https://github.com/10XGenomics/bamtofastq/releases/tag/v1.4.1. In case of LINUX systems you can use the following command when in cloned IDEIS directory:
wget https://github.com/10XGenomics/bamtofastq/releases/download/v1.4.1/bamtofastq_linux
mv bamtofastq_linux bamtofastq-1.4.1

# You might also need to make bamtofastq executable, which can be done this way:
chmod ugo+x bamtofastq-1.4.1

# example run
python IDEIS_main.py Ref_Human /media/edelaron/Midas/packages/ideis/Test_Data_IDEIS/possorted_genome_bam_excerpt.bam Results_data -g Homo-Sapiens --data-set /media/edelaron/Midas/packages/ideis/Test_Data_IDEIS/Test_data.rds

# got this issue:
## samtools: /home/edelaron/miniconda3/envs/IDEIS_env/bin/../lib/libtinfow.so.6: no version information available (required by samtools)
## samtools: /home/edelaron/miniconda3/envs/IDEIS_env/bin/../lib/libncursesw.so.6: no version information available (required by samtools)
## samtools: /home/edelaron/miniconda3/envs/IDEIS_env/bin/../lib/libncursesw.so.6: no version information available (required by samtools)
grep libtinfo /home/edelaron/miniconda3/envs/IDEIS_env/conda-meta/*
conda install samtools
conda install ncurses
# that fixed it, thanks github.com/bioconda/bioconda-recipes/issues/8210

# to see all options
python IDEIS_main.py -h
```

Test the example run:

```{r ideis-example-umap}
#| fig.dim = c(12, 6)

# or wherever you saved it
seu.data <- readRDS(file.path("~", "git", "Papers", "IDEIS", "Results_data",
                              "Results", "Datasets", "dataset_with_cd45.rds"))
seu.data <- NormalizeData(seu.data, assay = "CD45isoforms",
                          normalization.method = "CLR")

FeaturePlot(seu.data, c("PTPRC-RA", "PTPRC-RO")) & clean_umap
## ...found in 'CD45isoforms' assay instead
```

## CD45er

[GitHub](https://github.com/getzlab/10x-cd45-isoform-quantification)

> "CD45er outputs probabilities for each read aligning to CD45 to belong to one of the isoforms, RAX, RABX, RBX, RBCX, RABCX, or RX. The total probability of CD45RA was calculated as the sum of probabilities of reads belonging to isoforms RAX, RABX, and RABCX. The probability of CD45RO corresponds to the value marked in the column RX." - IDEIS paper

```{bash}
#| class.source = "fold-show",
#| eval = FALSE

cd git/Papers/
git clone https://github.com/getzlab/10x-cd45-isoform-quantification.git
cd 10x-cd45-isoform-quantification

python ./CD45er/run.py /media/edelaron/Midas/lyme/dataset2/default/192567SKL_GEX_HHT/outs/possorted_genome_bam.bam reference/CD45_exons.hg38.txt

<!-- Filtering 13192/24185 reads with minimum possible length >=500 -->
<!-- 5647/24185 reads informative for splicing -->
<!-- Estimated parameters: -->
<!-- mu:298.02400355632017,var:15546.585700865076 -->
<!-- RX       0.484523 -->
<!-- RAX      0.000040 -->
<!-- RBX      0.199155 -->
<!-- RABX     0.092181 -->
<!-- RBCX     0.096348 -->
<!-- RABCX    0.127753 -->
<!-- dtype: float64 -->

python ./CD45er/run.py /media/edelaron/Midas/lyme/dataset2/default/192567SKN_GEX_HHT/outs/possorted_genome_bam.bam reference/CD45_exons.hg38.txt

<!-- Filtering 6583/11024 reads with minimum possible length >=500 -->
<!-- 2732/11024 reads informative for splicing -->
<!-- Estimated parameters: -->
<!-- mu:302.89669342517647,var:17467.920803797428 -->
<!-- RX       0.568801 -->
<!-- RAX      0.005764 -->
<!-- RBX      0.249378 -->
<!-- RABX     0.062134 -->
<!-- RBCX     0.070661 -->
<!-- RABCX    0.043262 -->
<!-- dtype: float64 -->
```

## Real data

I wrote a script:

* I only kept the `Cell_ID_Unique` column in the Seurat objects which won't match with the bam file cell ids, so we're not going to use the whitelist/Seurat object option.
* The default expected cells is 6000, but that's not what I saw in the QC so I changed it to 4000 instead (which didn't end up changing the total mapped reads)
* The full script runs in a couple of minutes, so I didn't add any more cores.

```{r raw-processing}
#| attr.output = 'style="max-height: 300px;"'

cat(readLines(file.path(path_base, "primary_analysis", "scripts",
                        "splicing.sh")), sep = "\n")
```

Parse the output:

```{r}
#| eval = FALSE

# setup
sample_names <- list.dirs(file.path(path_data_root, "splicing"),
                          full.names = FALSE, recursive = FALSE)
cell_id_names <- c()
cd45_objs <- list()

# exclude the blood data
sample_names <- meta %>%
                  filter(SampleDir %in% sample_names, TissueType == "Skin") %>%
                  pull(SampleDir)

# read in the spliced data for all of the samples
for (sample in sample_names) {
  files <- file.path(path_data_root, "splicing", sample, "Results", "alevin",
                     "alevin", "quants_mat.gz")

  if (file.exists(files)) {
    txi <- tximport(files, type = "alevin")
    cd45_objs[[sample]] <- CreateSeuratObject(counts = txi$counts)

    # perhaps not all of the samples ran through, so we can't just pull metadata columns
    cell_id_names <- c(cell_id_names, sample)
  }
}

# fixed and cleaned names
cell_id_names <-
  meta %>% filter(SampleDir %in% cell_id_names) %>% pull(SampleName)

# merge all of the results into one object
cd45_obj <- merge(cd45_objs[[1]], y = tail(cd45_objs, -1),
                  add.cell.ids = cell_id_names)
cd45_obj$Cell_ID_Unique <- Cells(cd45_obj)

# add the data to combined_skin_t
combined_skin_t_ideis <- combined_skin_t
combined_skin_t_ideis <-
  subset(combined_skin_t_ideis, Cell_ID_Unique %in% Cells(cd45_obj))
cd45_obj <-
  subset(cd45_obj, Cell_ID_Unique %in% Cells(combined_skin_t_ideis))

# don't replace the RNA assay
combined_skin_t_ideis[["CD45"]] <- cd45_obj@assays$RNA
## Warning message:
## Cannot add objects with duplicate keys (offending key: rna_), setting key to 'cd45_'
DefaultAssay(combined_skin_t_ideis) <- "CD45"
cd45_feats <- str_c("PTPRC-R", c("A", "B", "C", "O"))

# save the object
saveRDS(combined_skin_t_ideis,
        file.path(path_objs, "downstream", "splicing",
                  "combined_skin_t_ideis.rds"))
```

Counts before and after:

```{r}
table(combined_skin_t$SampleName)
table(combined_skin_t_ideis$SampleName)
table(combined_skin_t_ideis$annotated_subclusters)

cat(paste(ncol(combined_skin_t_ideis), "cells out of the",
          ncol(combined_skin_t), "total skin T cells had spliced reads mapped."))
```

Visualization:

```{r ideis-umap-dot}
#| fig.dim = c(14, 16)

p1 <- UMAPPlot(combined_skin_t_ideis, cols = colors_annotated_skin_t,
               pt.size = 0.2) +
        labs(title = "Skin T Cells",
             subtitle = paste(ncol(combined_skin_t_ideis), "cells only")) +
        labels_standard + clean_umap
p2 <- DotPlot(combined_skin_t_ideis, features = cd45_feats, dot.scale = 3) %>%
        dot_color_scale() %>%
        plot_dot_side(seurat_obj = combined_skin_t_ideis,
                      row_identity = "annotated_subclusters",
                      info_to_add = "cluster_size") +
        labs(title = "CD45 Isoforms") +
        labels_standard + theme(axis.title = element_blank())
p3 <- FeaturePlot(combined_skin_t_ideis, features = cd45_feats,
                  pt.size = 0.2, order = TRUE, ncol = length(cd45_feats),
                  raster = FALSE) &
        scale_color_viridis(option = "G", direction = -1) &
        labs(color = "Expression") &
        labels_standard & clean_umap

p4 <- list()
for (marker in cd45_feats) {
  p4[[marker]] <-
      plot_density(combined_skin_t_ideis, features = marker, size = 0.2,
                   raster = FALSE) + labels_standard + clean_umap
}
p4 <- wrap_plots(p4, ncol = length(cd45_feats))

p5 <- vln_cell_info(seurat_obj = combined_skin_t_ideis, feature = "PTPRC-RA",
                    clrs_specific = colors_annotated_skin_t,
                    split_by = NULL, pt_size = 0.2, num_dec = 1, label_size = 3)

((p1 | p2) / p3 / p4 / free(p5)) + plot_layout(heights = c(2, 1, 1, 3))
```
