---
title: "TRUST4"
output:
  html_document:
    code_download: yes
    code_folding: hide
    css: ../../"style.css"
    dev: "png"
    df_print: kable
    fig_caption: yes
    keep_md: no
    self_contained: true
    theme: yeti
    toc: yes
    toc_depth: 4
    toc_float:
        collapsed: false
        smooth_scroll: false
editor_options:
  chunk_output_type: console
---

------------------------------------------------------------------------

# Overview and setup

```{r setup-options, include = FALSE}
knitr::opts_chunk$set(comment = NA, dpi = 200, fig.align = "center",
                      out.width = "100%", warning = FALSE, error = TRUE,
                      message = FALSE,
                      fig.path = here::here("primary_analysis", "figures",
                                            "downstream",
                                            "trust4", "trust4-"))

# will mess with chunk labels, but needed for parameterized reports
options(knitr.duplicate.label = "allow")
```

**Author(s):** Edel Aron<br>
**Last Updated:** `r Sys.Date()`

**R version:** `r R.Version()$version.string`<br>
**Platform:** `r R.Version()$platform`<br>
**Running under:** `r sessionInfo()$running`

```{r, include = FALSE}
# load packages, colors and themes and functions (R code only)
code_setup <-
  knitr::purl(here::here("primary_analysis", "notebooks", "01-setup.Rmd"),
              output = tempfile(), documentation = 0)
code_functions <-
  knitr::purl(here::here("primary_analysis", "notebooks", "02-functions.Rmd"),
              output = tempfile(), documentation = 0)
code_meta <- 
  knitr::purl(here::here("primary_analysis", "notebooks", "03-meta.Rmd"),
              output = tempfile(), documentation = 0)

# read in the code, then remove the temp files
source(code_setup)
source(code_functions)
source(code_meta)
unlink(c(code_setup, code_functions, code_meta))
```

10x Genomics [doesn't support gamma-delta TCRs](https://kb.10xgenomics.com/hc/en-us/articles/360015793931-Can-I-detect-T-cells-with-gamma-delta-chains-in-my-V-D-J-data), so we can try to detect by reconstructing TCR data from GEX data using [TRUST4](https://www.nature.com/articles/s41592-021-01142-2). At the end of this notebook, we also plot some TRG and TRD genes.

*Note that the TRUST4 points (in red) are set to be larger than the TCR overlays (in green) so that they are easier to see.*

Thanks to Edward Lee in the Kleinstein lab for running TRUST4 for us!

------------------------------------------------------------------------

# Read in the data

## GEX

```{r setup-gex}
# objects
combined_blood <-
  readRDS(file.path(path_objs, "airr", "combined_blood_airr.rds"))
combined_skin <-
  readRDS(file.path(path_objs, "airr", "combined_skin_airr.rds"))
```

```{r setup-gex-round-1}
# objects
combined_blood_t_full <- readRDS(file.path(path_objs, "combined_blood_t.rds"))
combined_skin_t_full <- readRDS(file.path(path_objs, "combined_skin_t.rds"))
```

```{r setup-gex-round-2}
# objects
combined_blood_t <- readRDS(file.path(path_objs, "combined_blood_t_ref.rds"))
combined_skin_t <- readRDS(file.path(path_objs, "combined_skin_t_ref.rds"))
```

## TRUST4

```{r setup-trust4}
trust4_all <- read_tsv(file.path(path_tables, "trust4_all.tsv"),
                       show_col_types = FALSE) %>%
                mutate(Dataset = as.character(Dataset))
```

------------------------------------------------------------------------

# Run the tool

> "TRUST4 outputs several files. trust_raw.out, trust_final.out are the contigs and corresponding nucleotide weight. trust_annot.fa is in fasta format for the annotation of the consensus assembly. trust_cdr3.out reports the CDR1,2,3 and gene information for each consensus assemblies. And trust_report.tsv is a report file focusing on CDR3 and is compatible with other repertoire analysis tool such as VDJTools."
- [GitHub](https://github.com/liulab-dfci/TRUST4)

## TRUST4

Make sure that you have the TRUST4 GitHub repository cloned so that you can use their scripts and references (you'll need hg38_bcrtcr.fa and human_IMGT+C.fa).

```{r trust4-example-script}
# each dataset got their own script for simplicity
readLines(file.path(path_base, "primary_analysis", "scripts", "trust4",
                    "trust4_lyme_dataset2.sh")) %>%
  paste0(collapse = "\n") %>% cat

# example of the output files:
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_airr_align.tsv
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_airr.tsv
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_annot.fa
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_assembled_reads.fa
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_barcode_airr.tsv
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_barcode_report.tsv
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_cdr3.out
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_final.out
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_raw.out
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_report.tsv
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_toassemble_1.fq
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_toassemble_2.fq
## TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_toassemble_bc.fa
```

In addition to the bash scripts that processed the raw fastqs from all three datasets, Ed also included two scripts from the TRUST4 repo that try to eliminate some falsely called TCRs from the results. According to him, the usage would basically be:

```{bash, eval = FALSE}
# use the *_barcode_report.tsv file for each sample
# for example
python3 barcoderep-filter.py -b TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_barcode_report.tsv > TRUST_192561SKL_filtered.tsv

# use the output of the previous step and the *_annot.fa file, and use the perl script to convert to AIRR format
perl trust-airr.pl TRUST_192561SKL_filtered.tsv \
TRUST_192561SKL_GEX_HHT_S1_L001_R1_001_annot.fa \
--format barcoderep > TRUST_192561SKL_airr_filtered.tsv
```

I didn't end up using these scripts though.

## Read in the output

Each sample gets its own directory and only files ending with "...airr.tsv" should be used. The "...barcode_airr.tsv" files have not been filtered by TRUST4's scoring system. They are based upon the constructed TCR sequences for each possible barcode.

The `sequence_id` in the "...airr.tsv" files should contain the usual 10x barcodes (plus additional numbers that TRUST4 adds).

Read through the script outputs:

```{r setup-filter-files}
#| eval = FALSE

trust4_all <- c()

# parent_dir <- file.path(path_data_root, "trust4")
# dir_list <- list.dirs(parent_dir, recursive = FALSE)

# GEX only
for (dir in filter(meta, DataType == "GEX")$SampleDir) {
  # get a list of all files in the directory that end with "_airr.tsv"
  file_list <- list.files(file.path(path_data_root, "trust4", paste0(dir, "_fqs")),
                          pattern = "_airr.tsv$", full.names = TRUE)
  file <- file_list[!grepl("_barcode_airr.tsv$", file_list)]

  # read in each file and bind the rows together, then add metadata
  df <- read_tsv(file, show_col_types = FALSE) %>%
          mutate(SampleName = filter(meta, SampleDir == dir)$SampleName,
                 TissueType = filter(meta, SampleDir == dir)$TissueType,
                 Dataset = filter(meta, SampleDir == dir)$Dataset)

  trust4_all <- bind_rows(trust4_all, df)
}

write_tsv(trust4_all, file.path(path_tables, "trust4_all.tsv"))
```

Some people think that you should filter out everything with `consensus_count > 1`, but I couldn't find enough support for that so I didn't do that.

## Plots

```{r template-umap}
#| eval = FALSE

# needs: trust4_obj, trust4_tissue_type,
#        trust4_clusters, trust4_cell_types

# assumes that trust4_obj contains a "Has_TCR" column

# by cluster
Idents(trust4_obj) <- trust4_clusters
p1 <- UMAPPlot(trust4_obj, label = TRUE, label.size = 3,
               cells.highlight = trust4_barcodes, sizes.highlight = 0.8,
               raster = FALSE) +
        scale_color_manual(name = "TRUST4",
                           labels = c("All", "γδ T Cells"),
                           values = c("lightgray", "red")) +
        labs(title = trust4_tissue_type, subtitle = "By Cluster") +
        labels_standard + clean_umap

# by annotation
Idents(trust4_obj) <- trust4_cell_types
p2 <- UMAPPlot(trust4_obj, label = TRUE, label.size = 3,
               cells.highlight = trust4_barcodes, sizes.highlight = 0.8,
               raster = FALSE) +
        scale_color_manual(name = "TRUST4",
                           labels = c("All", "γδ T Cells"),
                           values = c("lightgray", "red")) +
        labs(title = trust4_tissue_type, subtitle = "By Cell Type") +
        labels_standard + clean_umap

# TCR overlay
p3 <- plot_immune_overlay(seurat_obj = trust4_obj,
                          tissue_type = trust4_tissue_type,
                          airr_type = "TCR", clrs_specific = colors_datatype,
                          plot_by = "all")

# combine plots
p1 | p2 | p3
```

```{r template-umap-alt}
#| eval = FALSE

# needs: trust4_obj, trust4_tissue_type,
#        trust4_clusters, trust4_cell_types

# assumes that trust4_obj contains a "Has_TCR" column

# by cluster
Idents(trust4_obj) <- trust4_clusters
p1 <- UMAPPlot(trust4_obj, label = TRUE, label.size = 3,
               cells.highlight = trust4_barcodes, sizes.highlight = 0.8,
               raster = FALSE) +
        scale_color_manual(name = "TRUST4",
                           labels = c("All", "γδ T Cells"),
                           values = c("lightgray", "red")) +
        labs(title = trust4_tissue_type, subtitle = "By Cluster") +
        labels_standard + clean_umap

# by annotation
Idents(trust4_obj) <- trust4_cell_types
p2 <- UMAPPlot(trust4_obj, label = TRUE, label.size = 3,
               cells.highlight = trust4_barcodes, sizes.highlight = 0.8,
               raster = FALSE) +
        scale_color_manual(name = "TRUST4",
                           labels = c("All", "γδ T Cells"),
                           values = c("lightgray", "red")) +
        labs(title = trust4_tissue_type, subtitle = "By Cell Type") +
        labels_standard + clean_umap

# assumes that trust4_obj contains a "Has_TCR" column
p3 <- UMAPPlot(trust4_obj,
               cells.highlight = Cells(subset(trust4_obj,
                                              subset = Has_TCR == FALSE)),
               sizes.highlight = 0.2, raster = FALSE) +
        scale_color_manual(name = "αβ TCR Presence",
                           labels = c("TCRs", "No TCRs"),
                           values = c(colors_datatype[["TCR"]], "black")) +
        labs(title = trust4_tissue_type) +
        labels_standard + clean_umap

# combine plots
p1 | p2 | p3
```

```{r template-bar}
#| eval = FALSE

# needs: trust4_obj, trust4_tissue_type, trust4_clusters, trust4_cell_types

# clusters
p4 <- ggplot(trust4_obj[[]] %>%
               dplyr::filter(Cell_ID_Unique %in% trust4_barcodes) %>%
               select(trust4_clusters, "Has_TCR") %>%
               mutate(Has_TCR = as.character(Has_TCR)),
             aes(x = !!sym(trust4_clusters), fill = Has_TCR)) +
        geom_bar(position = position_dodge2(preserve = "single"),
                 color = "black", linewidth = 0.2, width = 1) +
        labs(title = paste0("TRUST4 γδ T Cells by Cluster (",
                            trust4_tissue_type, ")"),
             x = "Cluster", y = "Count") +
        scale_y_continuous(breaks = breaks_pretty()) +
        scale_fill_manual(values = colors_binary) +
        theme_bw + labels_standard

# cell types
p5 <- ggplot(trust4_obj[[]] %>%
               dplyr::filter(Cell_ID_Unique %in% trust4_barcodes) %>%
               select(trust4_cell_types, "Has_TCR") %>%
               mutate(Has_TCR = as.character(Has_TCR)),
             aes(x = !!sym(trust4_cell_types), fill = Has_TCR)) +
        geom_bar(position = position_dodge2(preserve = "single"),
                 color = "black", linewidth = 0.2, width = 1) +
        labs(title = paste0("TRUST4 γδ T Cells by Cell Type (",
                            trust4_tissue_type, ")"),
             x = "Cell Type", y = "Count") +
        scale_y_continuous(breaks = breaks_pretty()) +
        scale_fill_manual(values = colors_binary) +
        theme_bw + labels_standard # + labels_rotate_x

# combine plots
p4 / p5 + plot_layout(guides = "collect")
```

```{r template-expanded-bar-clusters}
#| eval = FALSE

# needs: trust4_obj, trust4_tissue_type, trust4_clusters

trust4_tcrs_full <- trust4_obj[[]] %>%
                      dplyr::filter(Cell_ID_Unique %in% trust4_barcodes) %>%
                      select(trust4_clusters, Has_TCR) %>%
                      mutate(Has_TCR = as.character(Has_TCR)) %>%
                      mutate(Has_TCR = str_replace(Has_TCR,
                                                   "TRUE", "TRUE_γδ")) %>%
                      mutate(Has_TCR = str_replace(Has_TCR,
                                                   "FALSE", "FALSE_γδ")) %>%
                      group_by(Has_TCR, !!sym(trust4_clusters)) %>%
                      summarize(n = n(), .groups = "drop")

trust4_tcrs <- trust4_obj[[]] %>%
                 select(trust4_clusters, Has_TCR) %>%
                 mutate(Has_TCR = as.character(Has_TCR)) %>%
                 group_by(Has_TCR, !!sym(trust4_clusters)) %>%
                 summarize(n = n(), .groups = "drop")

trust4_tcrs_full <- bind_rows(trust4_tcrs_full, trust4_tcrs)

# stacked bar plot
p6 <- ggplot(trust4_tcrs_full,
             aes(x = !!sym(trust4_clusters), fill = Has_TCR, weight = n)) +
        geom_bar(position = "fill", color = "black", linewidth = 0.2) +
        labs(title = paste0("TCRs by Cluster (", trust4_tissue_type, ")"),
             subtitle = "TRUST4 γδ T Cells Included",
             x = "Cluster", y = "Fraction") +
        scale_y_continuous(labels = percent, expand = expansion(mult = .02)) +
        scale_fill_manual(values = c("FALSE" = colors_binary[["FALSE"]],
                                     "FALSE_γδ" = "#61553b",
                                     "TRUE" = colors_binary[["TRUE"]],
                                     "TRUE_γδ" = "#107a63")) +
        theme_bw + labels_standard

# counts
p7 <- ggplot(trust4_tcrs_full,
             aes(x = !!sym(trust4_clusters), fill = Has_TCR, weight = n)) +
        geom_bar(position = "dodge", color = "black", linewidth = 0.2) +
        labs(title = paste0("TCRs by Cluster (", trust4_tissue_type, ")"),
             subtitle = "TRUST4 γδ T Cells Included",
             x = "Cluster", y = "Count") +
        scale_fill_manual(values = c("FALSE" = colors_binary[["FALSE"]],
                                     "FALSE_γδ" = "#61553b",
                                     "TRUE" = colors_binary[["TRUE"]],
                                     "TRUE_γδ" = "#107a63")) +
        theme_bw + labels_standard

# combine plots
p6 / p7 + plot_layout(guides = "collect")
```

```{r template-expanded-bar-cell-types}
#| eval = FALSE

# needs: trust4_obj, trust4_tissue_type, trust4_cell_types

trust4_tcrs_full <- trust4_obj[[]] %>%
                      dplyr::filter(Cell_ID_Unique %in% trust4_barcodes) %>%
                      select(trust4_cell_types, Has_TCR) %>%
                      mutate(Has_TCR = as.character(Has_TCR)) %>%
                      mutate(Has_TCR = str_replace(Has_TCR,
                                                   "TRUE", "TRUE_γδ")) %>%
                      mutate(Has_TCR = str_replace(Has_TCR,
                                                   "FALSE", "FALSE_γδ")) %>%
                      group_by(Has_TCR, !!sym(trust4_cell_types)) %>%
                      summarize(n = n(), .groups = "drop")

trust4_tcrs <- trust4_obj[[]] %>%
                 select(trust4_cell_types, Has_TCR) %>%
                 mutate(Has_TCR = as.character(Has_TCR)) %>%
                 group_by(Has_TCR, !!sym(trust4_cell_types)) %>%
                 summarize(n = n(), .groups = "drop")

trust4_tcrs_full <- bind_rows(trust4_tcrs_full, trust4_tcrs)

# stacked bar plot
p6 <- ggplot(trust4_tcrs_full,
             aes(x = !!sym(trust4_cell_types), fill = Has_TCR, weight = n)) +
        geom_bar(position = "fill", color = "black", linewidth = 0.2) +
        labs(title = paste0("TCRs by Cluster (", trust4_tissue_type, ")"),
             subtitle = "TRUST4 γδ T Cells Included",
             x = "Cluster", y = "Fraction") +
        scale_y_continuous(labels = percent, expand = expansion(mult = .02)) +
        scale_fill_manual(values = c("FALSE" = colors_binary[["FALSE"]],
                                     "FALSE_γδ" = "#61553b",
                                     "TRUE" = colors_binary[["TRUE"]],
                                     "TRUE_γδ" = "#107a63")) +
        theme_bw + labels_standard

# counts
p7 <- ggplot(trust4_tcrs_full,
             aes(x = !!sym(trust4_cell_types), fill = Has_TCR, weight = n)) +
        geom_bar(position = "dodge", color = "black", linewidth = 0.2) +
        labs(title = paste0("TCRs by Cluster (", trust4_tissue_type, ")"),
             subtitle = "TRUST4 γδ T Cells Included",
             x = "Cluster", y = "Count") +
        scale_fill_manual(values = c("FALSE" = colors_binary[["FALSE"]],
                                     "FALSE_γδ" = "#61553b",
                                     "TRUE" = colors_binary[["TRUE"]],
                                     "TRUE_γδ" = "#107a63")) +
        theme_bw + labels_standard

# combine plots
p6 / p7 + plot_layout(guides = "collect")
```

------------------------------------------------------------------------

# Results

If we don't filter for `consensus_count > 1`:

## Filter the data

```{r filters}
trust4 <- trust4_all
paste("Original TRUST4 results:", nrow(trust4))

# just the gamma-delta T cells
trust4 <- trust4 %>% dplyr::filter(grepl("TRD|TRG", v_call))
paste("Just γδ T Cells:", nrow(trust4))

# for the objects
trust4_barcodes_full <- trust4 %>%
                          select(SampleName, sequence_id) %>%
                          mutate(sequence_id = sapply(strsplit(sequence_id, "_"),
                                                      function(x) x[[1]])) %>%
                          mutate(Cell_ID_Unique = paste0(SampleName, "_",
                                                         sequence_id),
                                TRUST4 = TRUE) %>%
                          select(-SampleName, -sequence_id)

# for the UMAPs
trust4_barcodes <- trust4_barcodes_full %>% pull(Cell_ID_Unique)
```

Examine the results:

```{r filters-table}
# all hits
print_kable(trust4)
```

## Counts

Look at the total TRUST4 hits by tissue type:

```{r bar}
#| fig.dim = c(12, 8)

# all tissues
p1 <- ggplot(trust4,
             aes(x = SampleName, fill = SampleName)) +
        geom_bar(color = "black", linewidth = 0.2) +
        labs(title = "TRUST4 γδ T Cells (All Tissues)",
             subtitle = "No consensus_count filtration",
             x = "Sample", y = "Count") +
        scale_y_continuous(breaks = breaks_pretty()) +
        scale_fill_manual(values = colors_samples) +
        facet_grid(~Dataset, scales = "free_x", space = "free",
                   labeller = purrr::partial(label_both, sep = " "))

# just blood
p2 <- ggplot(dplyr::filter(trust4, TissueType == "Blood"),
             aes(x = SampleName, fill = SampleName)) +
        geom_bar(color = "black", linewidth = 0.2) +
        labs(title = "TRUST4 γδ T Cells (Blood)",
             subtitle = "No consensus_count filtration",
             x = "Sample", y = "Count") +
        scale_y_continuous(breaks = breaks_pretty()) +
        scale_fill_manual(values = colors_sample_blood) +
        facet_grid(~Dataset, scales = "free_x", space = "free",
                   labeller = purrr::partial(label_both, sep = " "))

# just skin
p3 <- ggplot(dplyr::filter(trust4, TissueType == "Skin"),
             aes(x = SampleName, fill = SampleName)) +
        geom_bar(color = "black", linewidth = 0.2) +
        labs(title = "TRUST4 γδ T Cells (Skin)",
             subtitle = "No consensus_count filtration",
             x = "Sample", y = "Count") +
        scale_y_continuous(breaks = breaks_pretty()) +
        scale_fill_manual(values = colors_sample_skin)

# all plots
(p1 | (p2 / p3)) &
  theme_bw & labels_standard & labels_rotate_x &
  theme(legend.position = "none") &
  facet_grid(~Dataset, scales = "free_x", space = "free",
             labeller = purrr::partial(label_both, sep = " "))
```

## Exploration {.tabset}

I'm going to look at all cells here. You could do something like filtering to cells without paired TCRs.

### Blood {.tabset}

```{r blood-setup}
trust4_obj <- combined_blood
trust4_tissue_type <- "Blood"
trust4_clusters <- "seurat_clusters"
trust4_cell_types <- "annotated_clusters"
```

```{r blood-hits}
num_hits <- nrow(dplyr::filter(trust4_obj[[]],
                               Cell_ID_Unique %in% trust4_barcodes))
cat("There are", num_hits, "TRUST4 hits out of", ncol(trust4_obj),
    "total cells",
    paste0("(", label_percent(accuracy = 0.01)(num_hits/ncol(trust4_obj)), ")"))
```

#### UMAPs

```{r blood-umap}
#| ref.label = "template-umap",
#| fig.dim = c(18, 6)
```

#### TRUST4 Hits Only

```{r blood-bar}
#| ref.label = "template-bar",
#| fig.dim = c(12, 6)
```

#### Full TRUST Hits

```{r blood-expanded-clusters-bar}
#| ref.label = "template-expanded-bar-clusters",
#| fig.dim = c(12, 10)
```

```{r blood-expanded-cell-types-bar}
#| ref.label = "template-expanded-bar-cell-types",
#| fig.dim = c(12, 10)
```

### Blood T Cells {.tabset}

```{r blood-t-full-setup}
trust4_obj <- combined_blood_t_full
trust4_tissue_type <- "Blood T Cells"
trust4_clusters <- "seurat_subclusters"
trust4_cell_types <- "annotated_subclusters"
```

```{r blood-t-full-hits}
num_hits <- nrow(dplyr::filter(trust4_obj[[]],
                               Cell_ID_Unique %in% trust4_barcodes))
cat("There are", num_hits, "TRUST4 hits out of", ncol(trust4_obj),
    "total cells",
    paste0("(", label_percent(accuracy = 0.01)(num_hits/ncol(trust4_obj)), ")"))
```

#### UMAPs

```{r blood-t-full-umap}
#| ref.label = "template-umap",
#| fig.dim = c(18, 6)
```

#### TRUST4 Hits Only

```{r blood-t-full-bar}
#| ref.label = "template-bar",
#| fig.dim = c(12, 6)
```

#### Full TRUST Hits

```{r blood-t-full-expanded-clusters-bar}
#| ref.label = "template-expanded-bar-clusters",
#| fig.dim = c(12, 10)
```

```{r blood-t-full-expanded-cell-types-bar}
#| ref.label = "template-expanded-bar-cell-types",
#| fig.dim = c(12, 10)
```

### Blood T Cells (Refined) {.tabset}

```{r blood-t-setup}
trust4_obj <- combined_blood_t
trust4_tissue_type <- "Blood T Cells (Refined)"
trust4_clusters <- "seurat_subclusters"
trust4_cell_types <- "annotated_subclusters"
```

```{r blood-t-hits}
num_hits <- nrow(dplyr::filter(trust4_obj[[]],
                               Cell_ID_Unique %in% trust4_barcodes))
cat("There are", num_hits, "TRUST4 hits out of", ncol(trust4_obj),
    "total cells",
    paste0("(", label_percent(accuracy = 0.01)(num_hits/ncol(trust4_obj)), ")"))
```

#### UMAPs

```{r blood-t-umap}
#| ref.label = "template-umap",
#| fig.dim = c(18, 6)
```

#### TRUST4 Hits Only

```{r blood-t-bar}
#| ref.label = "template-bar",
#| fig.dim = c(12, 6)
```

#### Full TRUST Hits

```{r blood-t-expanded-clusters-bar}
#| ref.label = "template-expanded-bar-clusters",
#| fig.dim = c(12, 10)
```

```{r blood-t-expanded-cell-types-bar}
#| ref.label = "template-expanded-bar-cell-types",
#| fig.dim = c(12, 10)
```

### Skin {.tabset}

```{r skin-setup}
trust4_obj <- combined_skin
trust4_tissue_type <- "Skin"
trust4_clusters <- "seurat_clusters"
trust4_cell_types <- "annotated_clusters"
```

```{r skin-hits}
num_hits <- nrow(dplyr::filter(trust4_obj[[]],
                               Cell_ID_Unique %in% trust4_barcodes))
cat("There are", num_hits, "TRUST4 hits out of", ncol(trust4_obj),
    "total cells",
    paste0("(", label_percent(accuracy = 0.01)(num_hits/ncol(trust4_obj)), ")"))
```

#### UMAPs

```{r skin-umap}
#| ref.label = "template-umap",
#| fig.dim = c(18, 6)
```

#### TRUST4 Hits Only

```{r skin-bar}
#| ref.label = "template-bar",
#| fig.dim = c(12, 6)
```

#### Full TRUST Hits

```{r skin-expanded-clusters-bar}
#| ref.label = "template-expanded-bar-clusters",
#| fig.dim = c(12, 10)
```

```{r skin-expanded-cell-types-bar}
#| ref.label = "template-expanded-bar-cell-types",
#| fig.dim = c(12, 10)
```

### Skin T Cells {.tabset}

```{r skin-t-full-setup}
trust4_obj <- combined_skin_t_full
trust4_tissue_type <- "Skin T Cells"
trust4_clusters <- "seurat_subclusters"
trust4_cell_types <- "annotated_subclusters"
```

```{r skin-t-full-hits}
num_hits <- nrow(dplyr::filter(trust4_obj[[]],
                               Cell_ID_Unique %in% trust4_barcodes))
cat("There are", num_hits, "TRUST4 hits out of", ncol(trust4_obj),
    "total cells",
    paste0("(", label_percent(accuracy = 0.01)(num_hits/ncol(trust4_obj)), ")"))
```

#### UMAPs

```{r skin-t-full-umap}
#| ref.label = "template-umap",
#| fig.dim = c(18, 6)
```

#### TRUST4 Hits Only

```{r skin-t-full-bar}
#| ref.label = "template-bar",
#| fig.dim = c(12, 6)
```

#### Full TRUST Hits

```{r skin-t-full-expanded-clusters-bar}
#| ref.label = "template-expanded-bar-clusters",
#| fig.dim = c(12, 10)
```

```{r skin-t-full-expanded-cell-types-bar}
#| ref.label = "template-expanded-bar-cell-types",
#| fig.dim = c(12, 10)
```

### Skin T Cells (Refined) {.tabset}

```{r skin-t-setup}
trust4_obj <- combined_skin_t
trust4_tissue_type <- "Skin T Cells (Refined)"
trust4_clusters <- "seurat_subclusters"
trust4_cell_types <- "annotated_clusters"
```

```{r skin-t-hits}
num_hits <- nrow(dplyr::filter(trust4_obj[[]],
                               Cell_ID_Unique %in% trust4_barcodes))
cat("There are", num_hits, "TRUST4 hits out of", ncol(trust4_obj),
    "total cells",
    paste0("(", label_percent(accuracy = 0.01)(num_hits/ncol(trust4_obj)), ")"))
```

#### UMAPs

```{r skin-t-umap}
#| ref.label = "template-umap",
#| fig.dim = c(18, 6)
```

#### TRUST4 Hits Only

```{r skin-t-bar}
#| ref.label = "template-bar",
#| fig.dim = c(12, 6)
```

#### Full TRUST Hits

```{r skin-t-expanded-clusters-bar}
#| ref.label = "template-expanded-bar-clusters",
#| fig.dim = c(12, 10)
```

```{r skin-t-expanded-cell-types-bar}
#| ref.label = "template-expanded-bar-cell-types",
#| fig.dim = c(12, 10)
```

------------------------------------------------------------------------

# Further exploration

## Blood

### Blood T Cells

It seems like the unrefined blood T cells might have a γδ T cell cluster.

```{r exploration-blood-t-full-setup}
trust4_obj <- combined_blood_t_full
trust4_tissue_type <- "Blood T Cells"
trust4_clusters <- "seurat_subclusters"
trust4_cell_types <- "annotated_subclusters"
```

```{r exploration-blood-t-full-umap}
#| ref.label = "template-umap-alt",
#| fig.dim = c(18, 6)
```

Let's check some good G and D genes (they are plotted on top for visibility):

```{r exploration-blood-t-full-feature}
#| fig.dim = c(12, 6)

FeaturePlot(object = trust4_obj, features = c("TRGV9", "TRDV2"),
            pt.size = 0.8, order = TRUE) &
  labels_standard & clean_umap
```

Our 4% γδ T cells in the blood fell within the expected range according to the TRUST4 authors (0.5 - 5%).

### Blood T Cells (Refined)

That cluster shows up in the refined cells as well.

```{r exploration-blood-t-setup}
trust4_obj <- combined_blood_t
trust4_tissue_type <- "Blood T Cells (Refined)"
trust4_clusters <- "seurat_subclusters"
trust4_cell_types <- "annotated_subclusters"
```

```{r exploration-blood-t-umap}
#| ref.label = "template-umap-alt",
#| fig.dim = c(18, 6)
```

Let's check some good G and D genes (they are plotted on top for visibility):

```{r exploration-blood-t-feature}
#| fig.dim = c(18, 6)

FeaturePlot(object = trust4_obj, features = c("TRGV9", "TRGV10", "TRDV2"),
            pt.size = 0.8, ncol = 3, order = TRUE) &
  labels_standard & clean_umap
```

## Skin

There's nothing quite as clear for the refined skin T cells:

```{r exploration-skin-t-feature}
#| fig.dim = c(12, 6)

FeaturePlot(object = combined_skin_t, features = c("TRGV9", "TRDV2"),
            pt.size = 0.8, order = TRUE) &
  labels_standard & clean_umap
```
